<!doctype html>
<html class="no-js">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

<meta property="og:title" content="References" />
  
<meta property="og:type" content="website" />
  
<meta property="og:url" content="https://ehrapy.readthedocs.io/en/latest/references.html" />
  
<meta property="og:description" content="Amir13, Amir et al.(2013), viSNE enables visualization of high dimensional single-cell data and reveals phenotypic heterogeneity of leukemia, Nature Biotechnology. Blondel08, Blondel et al.(2008), ..." />
  
<meta property="og:image" content="https://ehrapy.readthedocs.io/en/latest//_static/logo.png" />
  
<meta property="og:image:alt" content="References" />
  <link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="prev" title="Credits" href="authors.html" />

    <meta name="generator" content="sphinx-4.5.0, furo 2022.04.07"/>
        <title>References - ehrapy</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=68f4518137b9aefe99b631505a2064c3c42c9852" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    <link rel="stylesheet" type="text/css" href="_static/css/override.css" />
    <link rel="stylesheet" type="text/css" href="_static/css/sphinx_gallery.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">ehrapy</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand centered" href="index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="_static/ehrapy_pure.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage/usage.html">Usage</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="tutorials/index.html">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/ehrapy_introduction.html">ehrapy Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/ehrapy_introduction.html#Fundamental-Principles">Fundamental Principles</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html">MIMIC-II IAC Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Preprocessing">Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#PCA-&amp;-UMAP">PCA &amp; UMAP</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Batch-effect-check">Batch effect check</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Leiden-clustering">Leiden clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Hospital-&amp;-ICU-statistics">Hospital &amp; ICU statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Death">Death</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Demographics">Demographics</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Comorbidities">Comorbidities</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Initial-lab-measurements">Initial lab measurements</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Annotation">Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_introduction.html#Paga">Paga</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_fate.html">MIMIC-II Patient Fate</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_fate.html#Determining-patient-fate-using-a-ConnectivityKernel">Determining patient fate using a ConnectivityKernel</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/mimic_2_fate.html#Determining-patient-fate-with-a-PseudotimeKernel">Determining patient fate with a PseudotimeKernel</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/diabetes_130.html">Diabetes 130 Vignette</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/diabetes_130.html#Visual-Exploration">Visual Exploration</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/medcat.html">Extracting from free text with MedCAT</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/medcat.html#Custom-MedCAT-object-of-ehrapy">Custom MedCAT object of ehrapy</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/notebooks/medcat.html#Extracting-and-visualizing-all-disease-entities">Extracting and visualizing all disease entities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributor Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_of_conduct.html">Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/theislab/ehrapy/discussions">Discussions</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">References</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container"><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="references">
<h1>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h1>
<dl class="citation">
<dt class="label" id="amir13"><span class="brackets">Amir13</span></dt>
<dd><p>Amir <em>et al.</em> (2013),
<em>viSNE enables visualization of high dimensional single-cell data and reveals phenotypic heterogeneity of leukemia</em>,
<a class="reference external" href="https://doi.org/10.1038/nbt.2594">Nature Biotechnology</a>.</p>
</dd>
<dt class="label" id="blondel08"><span class="brackets">Blondel08</span></dt>
<dd><p>Blondel <em>et al.</em> (2008),
<em>Fast unfolding of communities in large networks</em>,
<a class="reference external" href="https://doi.org/10.1088/1742-5468/2008/10/P10008">J. Stat. Mech.</a>.</p>
</dd>
<dt class="label" id="chippada18"><span class="brackets">Chippada18</span></dt>
<dd><p><em>ForceAtlas2 for Python and NetworkX</em>,
<a class="reference external" href="https://github.com/bhargavchippada/forceatlas2">GitHub</a>.</p>
</dd>
<dt class="label" id="coifman05"><span class="brackets">Coifman05</span></dt>
<dd><p>Coifman <em>et al.</em> (2005),
<em>Geometric diffusions as a tool for harmonic analysis and structure definition of data: Diffusion maps</em>,
<a class="reference external" href="https://doi.org/10.1038/nmeth.3971">PNAS</a>.</p>
</dd>
<dt class="label" id="csardi06"><span class="brackets">Csardi06</span></dt>
<dd><p>Csardi <em>et al.</em> (2006),
<em>The igraph software package for complex network research</em>,
<a class="reference external" href="http://igraph.org">InterJournal Complex Systems</a>.</p>
</dd>
<dt class="label" id="fruchterman91"><span class="brackets">Fruchterman91</span></dt>
<dd><p>Fruchterman &amp; Reingold (1991),
<em>Graph drawing by force-directed placement</em>,
<a class="reference external" href="http://doi.org:10.1002/spe.4380211102">Software: Practice &amp; Experience</a>.</p>
</dd>
<dt class="label" id="haghverdi15"><span class="brackets">Haghverdi15</span></dt>
<dd><p>Haghverdi <em>et al.</em> (2015),
<em>Diffusion maps for high-dimensional single-cell analysis of differentiation data</em>,
<a class="reference external" href="https://doi.org/10.1093/bioinformatics/btv325">Bioinformatics</a>.</p>
</dd>
<dt class="label" id="haghverdi16"><span class="brackets">Haghverdi16</span></dt>
<dd><p>Haghverdi <em>et al.</em> (2016),
<em>Diffusion pseudotime robustly reconstructs branching cellular lineages</em>,
<a class="reference external" href="https://doi.org/10.1038/nmeth.3971">Nature Methods</a>.</p>
</dd>
<dt class="label" id="islam11"><span class="brackets">Islam11</span></dt>
<dd><p>Islam <em>et al.</em> (2011),
<em>Characterization of the single-cell transcriptional landscape by highly multiplex RNA-seq</em>,
<a class="reference external" href="https://doi.org/10.1101/gr.110882.110">Genome Research</a>.</p>
</dd>
<dt class="label" id="jacomy14"><span class="brackets">Jacomy14</span></dt>
<dd><p>Jacomy <em>et al.</em> (2014),
<em>ForceAtlas2, a Continuous Graph Layout Algorithm for Handy Network Visualization Designed for the Gephi Software</em>
<a class="reference external" href="https://doi.org/10.1371/journal.pone.0098679">PLOS One</a>.</p>
</dd>
<dt class="label" id="johnson07"><span class="brackets">Johnson07</span></dt>
<dd><p>Johnson, Li &amp; Rabinovic (2007),
<em>Adjusting batch effects in microarray expression data using empirical Bayes methods</em>,
<a class="reference external" href="https://doi.org/10.1093/biostatistics/kxj037">Biostatistics</a>.</p>
</dd>
<dt class="label" id="lambiotte09"><span class="brackets">Lambiotte09</span></dt>
<dd><p>Lambiotte <em>et al.</em> (2009)
<em>Laplacian Dynamics and Multiscale Modular Structure in Networks</em>
<a class="reference external" href="https://arxiv.org/abs/0812.1770">arXiv</a>.</p>
</dd>
<dt class="label" id="leek12"><span class="brackets">Leek12</span></dt>
<dd><p>Leek <em>et al.</em> (2012),
<em>sva: Surrogate Variable Analysis. R package</em>
<a class="reference external" href="https://doi.org/10.18129/B9.bioc.sva">Bioconductor</a>.</p>
</dd>
<dt class="label" id="levine15"><span class="brackets">Levine15</span></dt>
<dd><p>Levine <em>et al.</em> (2015),
<em>Data-Driven Phenotypic Dissection of AML Reveals Progenitor–like Cells that Correlate with Prognosis</em>,
<a class="reference external" href="https://doi.org/10.1016/j.cell.2015.05.047">Cell</a>.</p>
</dd>
<dt class="label" id="maaten08"><span class="brackets">Maaten08</span></dt>
<dd><p>Maaten &amp; Hinton (2008),
<em>Visualizing data using t-SNE</em>,
<a class="reference external" href="http://www.jmlr.org/papers/v9/vandermaaten08a.html">JMLR</a>.</p>
</dd>
<dt class="label" id="mcinnes18"><span class="brackets">McInnes18</span></dt>
<dd><p>McInnes &amp; Healy (2018),
<em>UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction</em>,
<a class="reference external" href="https://arxiv.org/abs/1802.03426">arXiv</a>.</p>
</dd>
<dt class="label" id="pedersen12"><span class="brackets">Pedersen12</span></dt>
<dd><p>Pedersen (2012),
<em>Python implementation of ComBat</em>
<a class="reference external" href="https://github.com/brentp/combat.py">GitHub</a>.</p>
</dd>
<dt class="label" id="pedregosa11"><span class="brackets">Pedregosa11</span></dt>
<dd><p>Pedregosa <em>et al.</em> (2011),
<em>Scikit-learn: Machine Learning in Python</em>,
<a class="reference external" href="http://www.jmlr.org/papers/v12/pedregosa11a.html">JMLR</a>.</p>
</dd>
<dt class="label" id="traag17"><span class="brackets">Traag17</span></dt>
<dd><p>(2017),
<em>Louvain</em>,
<a class="reference external" href="https://doi.org/10.5281/zenodo.35117">GitHub</a>.</p>
</dd>
<dt class="label" id="traag18"><span class="brackets">Traag18</span></dt>
<dd><p>Traag <em>et al.</em> (2018),
<em>From Louvain to Leiden: guaranteeing well-connected communities</em>
<a class="reference external" href="https://arxiv.org/abs/1810.08473">arXiv</a>.</p>
</dd>
<dt class="label" id="weinreb17"><span class="brackets">Weinreb17</span></dt>
<dd><p>Weinreb <em>et al.</em> (2016),
<em>SPRING: a kinetic interface for visualizing high dimensional single-cell expression data</em>,
<a class="reference external" href="https://doi.org/10.1101/090332">bioRxiv</a>.</p>
</dd>
<dt class="label" id="wolf18"><span class="brackets">Wolf18</span></dt>
<dd><p>Wolf <em>et al.</em> (2018),
<em>Scanpy: large-scale single-cell gene expression data analysis</em>,
<a class="reference external" href="https://doi.org/10.1186/s13059-017-1382-0">Genome Biology</a>.</p>
</dd>
<dt class="label" id="wolf19"><span class="brackets">Wolf19</span></dt>
<dd><p>Wolf <em>et al.</em> (2019),
<em>PAGA: Graph abstraction reconciles clustering with trajectory inference through a topology preserving map of single cells.</em>
<a class="reference external" href="https://doi.org/10.1186/s13059-019-1663-x">Genome Biology</a>,
<a class="reference external" href="https://doi.org/10.1101/208819">bioRxiv</a>.</p>
</dd>
<dt class="label" id="zunder15"><span class="brackets">Zunder15</span></dt>
<dd><p>Zunder <em>et al.</em> (2015),
<em>A continuous molecular roadmap to iPSC reprogramming through progression analysis of single-cell mass cytometry</em>,
<a class="reference external" href="https://doi.org/10.1016/j.stem.2015.01.015">Cell Stem Cell</a>.</p>
</dd>
</dl>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="authors.html">
              <svg><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Credits</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021, Lukas Heumos, Theislab
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/scripts/furo.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/design-tabs.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>