
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>MIMIC-II Patient Fate &#8212; ehrapy 0.2.0 documentation</title>
    
    <link href="../../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
  
    
    <link rel="stylesheet"
      href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
    <link rel="preload" as="font" type="font/woff2" crossorigin
      href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
    <link rel="preload" as="font" type="font/woff2" crossorigin
      href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">
  
    
      
  
    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/pydata-sphinx-theme.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom_cookietemple.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/dataframe.css" />
    
    <link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">
  
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Contributor Guide" href="../../contributing.html" />
    <link rel="prev" title="MIMIC-II IAC Introduction" href="mimic_2_introduction.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    
    <nav class="navbar navbar-light navbar-expand-lg bg-light fixed-top bd-navbar" id="navbar-main"><div class="container-xl">

  <div id="navbar-start">
    
    

<a class="navbar-brand" href="../../index.html">
  <img src="../../_static/placeholder.png" class="logo" alt="logo">
</a>


    
  </div>

  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-collapsible" aria-controls="navbar-collapsible" aria-expanded="false" aria-label="Toggle navigation">
    <span class="navbar-toggler-icon"></span>
  </button>

  
  <div id="navbar-collapsible" class="col-lg-9 collapse navbar-collapse">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <ul id="navbar-main-elements" class="navbar-nav">
    <li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../../installation.html">
  Installation
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../../usage/usage.html">
  Usage
 </a>
</li>

<li class="toctree-l1 current active nav-item">
 <a class="reference internal nav-link" href="../index.html">
  Tutorials
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../../contributing.html">
  Contributor Guide
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../../code_of_conduct.html">
  Code of Conduct
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference external nav-link" href="https://github.com/theislab/ehrapy/discussions">
  Discussions
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../../authors.html">
  Credits
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../../references.html">
  References
 </a>
</li>

    
</ul>
      </div>
      
    </div>

    <div id="navbar-end">
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          <a class="nav-link" href="https://github.com/theislab/ehrapy" rel="noopener" target="_blank" title="GitHub"><span><i class="fab fa-github-square"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="https://twitter.com/lukasheumos" rel="noopener" target="_blank" title="Twitter"><span><i class="fab fa-twitter-square"></i></span>
            <label class="sr-only">Twitter</label></a>
        </li>
      </ul>
      </div>
      
    </div>
  </div>
</div>
    </nav>
    

    <div class="container-xl">
      <div class="row">
          
            
            <!-- Only show if we have sidebars configured, else just a small margin  -->
            <div class="col-12 col-md-3 bd-sidebar">
              <div class="sidebar-start-items"><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  <div class="bd-toc-item active">
    <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="ehrapy_introduction.html">
   ehrapy introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mimic_2_introduction.html">
   MIMIC-II IAC Introduction
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   MIMIC-II Patient Fate
  </a>
 </li>
</ul>

  </div>
</nav>
              </div>
              <div class="sidebar-end-items">
              </div>
            </div>
            
          

          
          <div class="d-none d-xl-block col-xl-2 bd-toc">
            
              
              <div class="toc-item">
                
<div class="tocsection onthispage pt-5 pb-3">
    <i class="fas fa-list"></i> On this page
</div>

<nav id="bd-toc-nav">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   MIMIC-II Patient Fate
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Importing-ehrapy-and-setting-plotting-parameters">
     Importing ehrapy and setting plotting parameters
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Getting-and-preprocessing-the-MIMIC-II-dataset">
     Getting and preprocessing the MIMIC-II dataset
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Determining-patient-fate-using-a-ConnectivityKernel">
   Determining patient fate using a ConnectivityKernel
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Connectivity-Kernels-and-stream-embeddings">
     Connectivity Kernels and stream embeddings
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Determining-macrostates-and-terminal-states">
     Determining macrostates and terminal states
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Calculating-absorption-probabilities">
     Calculating absorption probabilities
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Extracting-lineage-drivers">
       Extracting lineage drivers
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Exploring-cluster-fates">
     Exploring cluster fates
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#Determining-patient-fate-with-a-PseudotimeKernel">
   Determining patient fate with a PseudotimeKernel
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Pseudotime-Kernels-and-stream-embeddings">
     Pseudotime Kernels and stream embeddings
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Simulating-transitions-with-random-walks">
     Simulating transitions with random walks
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Determining macrostates and terminal states
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Calculating absorption probabilities
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     Extracting lineage drivers
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Determining-feature-trends">
     Determining feature trends
    </a>
   </li>
  </ul>
 </li>
</ul>

</nav>
              </div>
              
              <div class="toc-item">
                

<div class="tocsection editthispage">
    <a href="https://github.com/theislab/ehrapy/edit/development/docs/tutorials/notebooks/mimic_2_fate.ipynb">
        <i class="fas fa-pencil-alt"></i> Edit this page
    </a>
</div>

              </div>
              
            
          </div>
          

          
          
            
          
          <main class="col-12 col-md-9 col-xl-7 py-md-5 pl-md-5 pr-md-4 bd-content" role="main">
              
              <div>
                
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="binder-badge docutils container">
    <a class="reference external image-reference"
       href="https://mybinder.org/v2/gh/theislab/ehrapy/development?filepath=docs/source/tutorials/notebooks/mimic_2_fate.ipynb">
    <img alt="Launch binder" src="https://mybinder.org/badge_logo.svg" width="150px">
    </a>
</div><div class="section" id="MIMIC-II-Patient-Fate">
<h1>MIMIC-II Patient Fate<a class="headerlink" href="#MIMIC-II-Patient-Fate" title="Permalink to this headline">¶</a></h1>
<p>This tutorial focuses on the determination of patient fate. The goal is to detect terminal states and the corresponding origins based on pseudotime. Very rarely does the actual real time reflect the actual progression of disease. Imagine that 100 patients come in at day 45 and get measurements done. Some will be healthy still (or recovered), some at the onset of a specific disease and some at the height of it. What we are much more interested in is a continuous transition of states (e.g. healthy
to diseased to dead). This can be achieved by identifying source states (e.g. healthy) and then calculating pseudotime from this state. Based on markov chain modelling we uncover patient dynamics using <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/index.html">cellrank</a>. For more details please read the <a class="reference external" href="https://www.nature.com/articles/s41592-021-01346-6">Cellrank paper</a>.</p>
<p>In this tutorial we will be using cellrank to:</p>
<ol class="arabic simple">
<li><p>Compute initial &amp; terminal as well as intermediate macrostates of our patient landscape.</p></li>
<li><p>Infer fate proabilities towards the terminal states for each individual patient (visit).</p></li>
<li><p>Visualize feature trends along specific patient states while accounting for the continuous nature of fate determination.</p></li>
<li><p>Identify potential driver features for each identified trajectory.</p></li>
</ol>
<p>A skim of the extensive and well written <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/index.html">cellrank documentation</a> is recommended. Especially the <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/cellrank_basics.html">basic tutorial</a> chapter is useful. If you are not familiar with single-cell data do not be afraid and replace cells with patients and genes with features in your mind.</p>
<p>This tutorial requires <code class="docutils literal notranslate"><span class="pre">cellrank</span></code> and <code class="docutils literal notranslate"><span class="pre">scvelo</span></code> to be installed. These are not dependencies of ehrapy and must be installed separately:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install cellrank scvelo
</pre></div>
</div>
<p>This tutorial uses the MIMIC-II IAC dataset. It was created for the purpose of a case study in the book: Secondary Analysis of Electronic Health Records, published by Springer in 2016. In particular, the dataset was used throughout Chapter 16 (Data Analysis) by Raffa J. et al. to investigate the effectiveness of indwelling arterial catheters in hemodynamically stable patients with respiratory failure for mortality outcomes. The dataset is derived from MIMIC-II, the publicly-accessible critical
care database. It contains summary clinical data and outcomes for 1,776 patients.</p>
<p>More details on the dataset can be found here: <a class="reference external" href="https://physionet.org/content/mimic2-iaccd/1.0/">https://physionet.org/content/mimic2-iaccd/1.0/</a>.</p>
<p>This tutorial will eventually load a preprocessed version of the MIMIC-II dataset, but for now we will start from scratch.</p>
<div class="section" id="Importing-ehrapy-and-setting-plotting-parameters">
<h2>Importing ehrapy and setting plotting parameters<a class="headerlink" href="#Importing-ehrapy-and-setting-plotting-parameters" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import ehrapy.api as ep
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import cellrank as cr
from cellrank.tl.kernels import ConnectivityKernel
from cellrank.tl.kernels import PseudotimeKernel
import scvelo as scv
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="color: #800000; text-decoration-color: #800000; font-weight: bold">Unable to contact PyPI to check for the latest ehrapy version. Do you have an internet </span>
<span style="color: #800000; text-decoration-color: #800000; font-weight: bold">connection?</span>
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/medcat/cat.py:16: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import warnings
from matplotlib import MatplotlibDeprecationWarning

warnings.simplefilter(&quot;ignore&quot;, category=UserWarning)
warnings.simplefilter(&quot;ignore&quot;, category=FutureWarning)
warnings.simplefilter(&quot;ignore&quot;, category=DeprecationWarning)
warnings.simplefilter(&quot;ignore&quot;, category=MatplotlibDeprecationWarning)
</pre></div>
</div>
</div>
</div>
<div class="section" id="Getting-and-preprocessing-the-MIMIC-II-dataset">
<h2>Getting and preprocessing the MIMIC-II dataset<a class="headerlink" href="#Getting-and-preprocessing-the-MIMIC-II-dataset" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = ep.dt.mimic_2(encoded=True)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pp.knn_impute(adata)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 1776 × 46
    obs: &#39;service_unit&#39;, &#39;day_icu_intime&#39;
    uns: &#39;categoricals&#39;, &#39;original_values_categoricals&#39;, &#39;var_to_encoding&#39;
    layers: &#39;original&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pp.pca(adata)
ep.pp.neighbors(adata)
ep.tl.umap(adata)
ep.tl.leiden(adata, resolution=0.5, key_added=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pl.umap(adata, color=[&quot;leiden_0_5&quot;], title=&quot;Leiden 0.5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;service_unit&#39; as categorical
... storing &#39;day_icu_intime&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_11_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_11_1.png" />
</div>
</div>
</div>
</div>
<div class="section" id="Determining-patient-fate-using-a-ConnectivityKernel">
<h1>Determining patient fate using a ConnectivityKernel<a class="headerlink" href="#Determining-patient-fate-using-a-ConnectivityKernel" title="Permalink to this headline">¶</a></h1>
<div class="section" id="Connectivity-Kernels-and-stream-embeddings">
<h2>Connectivity Kernels and stream embeddings<a class="headerlink" href="#Connectivity-Kernels-and-stream-embeddings" title="Permalink to this headline">¶</a></h2>
<p>Depending on the data it may not always be possible to clearly define a cluster or specific patients as the origin. When working with single-cell data this is easier because when stem cells are detected these are usually the origin of cell differentiation processes.</p>
<p>In our case we do not have any clear origin and will therefore start with <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/classes.html#connectivity-kernel">cellrank’s ConnectivityKernel</a>. This kernel computes transition probabilities based on similarities among patients using a KNN graph.</p>
<p>First we define the kernel, compute the transition matrix and a projection ontop of the UMAP.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ck = ConnectivityKernel(adata)
ck.compute_transition_matrix()
ck.compute_projection(basis=&quot;umap&quot;)
</pre></div>
</div>
</div>
<p>Note that ConnectivityKernel has a <code class="docutils literal notranslate"><span class="pre">backward</span></code> parameter which can be set to <code class="docutils literal notranslate"><span class="pre">True</span></code> to compute everything backwards in time.</p>
<p>We can now visualize the project forwards and backwards in time using <a class="reference external" href="https://scvelo.readthedocs.io/">scvelo</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scv.pl.velocity_embedding_stream(adata, vkey=&quot;T_fwd&quot;, basis=&quot;umap&quot;, color=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_20_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_20_0.png" />
</div>
</div>
<p>We observe that several clusters including e.g. 8 only really have outgoing arrows indicating steady states.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pl.umap(adata, color=&quot;hosp_exp_flg&quot;, title=&quot;Died in Hospital&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_22_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_22_0.png" />
</div>
</div>
<p>Cluster 8 is apparently the cluster where most people died and clusters 2 and 7 further include many deceased patients.</p>
<p>Let’s try to find the origins of the “death clusters”. We will now define a <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/api/cellrank.tl.estimators.GPCCA.html">GPCCA</a> estimator to predict the patient fates using the above calculated transition matrix.</p>
</div>
<div class="section" id="Determining-macrostates-and-terminal-states">
<h2>Determining macrostates and terminal states<a class="headerlink" href="#Determining-macrostates-and-terminal-states" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g = cr.tl.estimators.GPCCA(ck)
</pre></div>
</div>
</div>
<p>As a first step we try to identify macrostates in the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_macrostates(n_states=10, cluster_key=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Unable to import `petsc4py` or `slepc4py`. Using `method=&#39;brandts&#39;`
WARNING: For `method=&#39;brandts&#39;`, dense matrix is required. Densifying
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.macrostates_memberships
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div style='scoped' class='rendered_html'><table class='dataframe'><tr><th style='background-color: #ffbb78; color: #000000; text-align: center; word-wrap: break-word; max-width: 100px'>10</th><th style='background-color: #8c564b; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>5</th><th style='background-color: #17becf; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>8</th><th style='background-color: #b5bd61; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>7</th><th style='background-color: #aec7e8; color: #000000; text-align: center; word-wrap: break-word; max-width: 100px'>9</th><th style='background-color: #279e68; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>2</th><th style='background-color: #d62728; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>3</th><th style='background-color: #aa40fc; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>4</th><th style='background-color: #ff7f0e; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>1</th><th style='background-color: #1f77b4; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>0</th></tr><tr><td style='text-align: right;'>0.001859</td><td style='text-align: right;'>0.037395</td><td style='text-align: right;'>0.532238</td><td style='text-align: right;'>0.124361</td><td style='text-align: right;'>0.045519</td><td style='text-align: right;'>0.050040</td><td style='text-align: right;'>0.119454</td><td style='text-align: right;'>0.075511</td><td style='text-align: right;'>0.004004</td><td style='text-align: right;'>0.009618</td></tr><tr><td style='text-align: right;'>0.000090</td><td style='text-align: right;'>0.060376</td><td style='text-align: right;'>0.043704</td><td style='text-align: right;'>0.063549</td><td style='text-align: right;'>0.010979</td><td style='text-align: right;'>0.022617</td><td style='text-align: right;'>0.156159</td><td style='text-align: right;'>0.021860</td><td style='text-align: right;'>0.429602</td><td style='text-align: right;'>0.191065</td></tr><tr><td style='text-align: right;'>0.008623</td><td style='text-align: right;'>0.081905</td><td style='text-align: right;'>0.015586</td><td style='text-align: right;'>0.032788</td><td style='text-align: right;'>0.020018</td><td style='text-align: right;'>0.003878</td><td style='text-align: right;'>0.398662</td><td style='text-align: right;'>0.388226</td><td style='text-align: right;'>0.016006</td><td style='text-align: right;'>0.034308</td></tr><tr><td style='text-align: right;'>0.012918</td><td style='text-align: right;'>0.115001</td><td style='text-align: right;'>0.439220</td><td style='text-align: right;'>0.012007</td><td style='text-align: right;'>0.092818</td><td style='text-align: right;'>0.065762</td><td style='text-align: right;'>0.196724</td><td style='text-align: right;'>0.027296</td><td style='text-align: right;'>0.023567</td><td style='text-align: right;'>0.014687</td></tr><tr><td style='text-align: right;'>0.013815</td><td style='text-align: right;'>0.055717</td><td style='text-align: right;'>0.040014</td><td style='text-align: right;'>0.029437</td><td style='text-align: right;'>0.060968</td><td style='text-align: right;'>0.009114</td><td style='text-align: right;'>0.555421</td><td style='text-align: right;'>0.174307</td><td style='text-align: right;'>0.037122</td><td style='text-align: right;'>0.024084</td></tr><tr><td style='text-align: right;'>0.004629</td><td style='text-align: right;'>0.070237</td><td style='text-align: right;'>0.046731</td><td style='text-align: right;'>0.038861</td><td style='text-align: right;'>0.039380</td><td style='text-align: right;'>0.006291</td><td style='text-align: right;'>0.247427</td><td style='text-align: right;'>0.473925</td><td style='text-align: right;'>0.040349</td><td style='text-align: right;'>0.032172</td></tr><tr><td style='text-align: right;'>0.000071</td><td style='text-align: right;'>0.058578</td><td style='text-align: right;'>0.043763</td><td style='text-align: right;'>0.072836</td><td style='text-align: right;'>0.013124</td><td style='text-align: right;'>0.035540</td><td style='text-align: right;'>0.151808</td><td style='text-align: right;'>0.025937</td><td style='text-align: right;'>0.436710</td><td style='text-align: right;'>0.161634</td></tr><tr><td style='text-align: right;'>0.000085</td><td style='text-align: right;'>0.029210</td><td style='text-align: right;'>0.074534</td><td style='text-align: right;'>0.041905</td><td style='text-align: right;'>0.061860</td><td style='text-align: right;'>0.014708</td><td style='text-align: right;'>0.087062</td><td style='text-align: right;'>0.128456</td><td style='text-align: right;'>0.510246</td><td style='text-align: right;'>0.051936</td></tr><tr><td style='text-align: right;'>0.000134</td><td style='text-align: right;'>0.045965</td><td style='text-align: right;'>0.059068</td><td style='text-align: right;'>0.053114</td><td style='text-align: right;'>0.035005</td><td style='text-align: right;'>0.015457</td><td style='text-align: right;'>0.124851</td><td style='text-align: right;'>0.069733</td><td style='text-align: right;'>0.489544</td><td style='text-align: right;'>0.107129</td></tr><tr><td style='text-align: right;'>0.044702</td><td style='text-align: right;'>0.278596</td><td style='text-align: right;'>0.296944</td><td style='text-align: right;'>0.010294</td><td style='text-align: right;'>0.118475</td><td style='text-align: right;'>0.050953</td><td style='text-align: right;'>0.113119</td><td style='text-align: right;'>0.041102</td><td style='text-align: right;'>0.017422</td><td style='text-align: right;'>0.028392</td></tr><tr><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td></tr><tr><td style='text-align: right;'>0.324482</td><td style='text-align: right;'>0.407008</td><td style='text-align: right;'>0.021723</td><td style='text-align: right;'>0.047119</td><td style='text-align: right;'>0.007673</td><td style='text-align: right;'>0.004914</td><td style='text-align: right;'>0.112574</td><td style='text-align: right;'>0.026008</td><td style='text-align: right;'>0.018635</td><td style='text-align: right;'>0.029866</td></tr><tr><td style='text-align: right;'>0.001595</td><td style='text-align: right;'>0.040637</td><td style='text-align: right;'>0.007877</td><td style='text-align: right;'>0.369916</td><td style='text-align: right;'>0.037631</td><td style='text-align: right;'>0.272886</td><td style='text-align: right;'>0.129972</td><td style='text-align: right;'>0.071193</td><td style='text-align: right;'>0.060553</td><td style='text-align: right;'>0.007741</td></tr><tr><td style='text-align: right;'>0.000399</td><td style='text-align: right;'>0.046466</td><td style='text-align: right;'>0.047693</td><td style='text-align: right;'>0.044299</td><td style='text-align: right;'>0.027626</td><td style='text-align: right;'>0.014593</td><td style='text-align: right;'>0.137795</td><td style='text-align: right;'>0.038300</td><td style='text-align: right;'>0.213935</td><td style='text-align: right;'>0.428893</td></tr><tr><td style='text-align: right;'>0.132281</td><td style='text-align: right;'>0.466817</td><td style='text-align: right;'>0.052621</td><td style='text-align: right;'>0.068024</td><td style='text-align: right;'>0.098778</td><td style='text-align: right;'>0.005206</td><td style='text-align: right;'>0.051394</td><td style='text-align: right;'>0.071338</td><td style='text-align: right;'>0.008867</td><td style='text-align: right;'>0.044674</td></tr><tr><td style='text-align: right;'>0.000193</td><td style='text-align: right;'>0.038756</td><td style='text-align: right;'>0.017385</td><td style='text-align: right;'>0.124865</td><td style='text-align: right;'>0.047935</td><td style='text-align: right;'>0.496199</td><td style='text-align: right;'>0.119621</td><td style='text-align: right;'>0.104455</td><td style='text-align: right;'>0.022338</td><td style='text-align: right;'>0.028254</td></tr><tr><td style='text-align: right;'>0.003257</td><td style='text-align: right;'>0.075469</td><td style='text-align: right;'>0.037850</td><td style='text-align: right;'>0.029760</td><td style='text-align: right;'>0.031026</td><td style='text-align: right;'>0.007500</td><td style='text-align: right;'>0.220407</td><td style='text-align: right;'>0.540873</td><td style='text-align: right;'>0.021722</td><td style='text-align: right;'>0.032136</td></tr><tr><td style='text-align: right;'>0.000155</td><td style='text-align: right;'>0.038509</td><td style='text-align: right;'>0.018660</td><td style='text-align: right;'>0.117994</td><td style='text-align: right;'>0.048277</td><td style='text-align: right;'>0.498824</td><td style='text-align: right;'>0.118957</td><td style='text-align: right;'>0.105388</td><td style='text-align: right;'>0.022588</td><td style='text-align: right;'>0.030649</td></tr><tr><td style='text-align: right;'>0.000503</td><td style='text-align: right;'>0.040400</td><td style='text-align: right;'>0.010820</td><td style='text-align: right;'>0.176052</td><td style='text-align: right;'>0.044830</td><td style='text-align: right;'>0.451900</td><td style='text-align: right;'>0.124419</td><td style='text-align: right;'>0.094994</td><td style='text-align: right;'>0.032065</td><td style='text-align: right;'>0.024017</td></tr><tr><td style='text-align: right;'>0.000276</td><td style='text-align: right;'>0.039231</td><td style='text-align: right;'>0.015925</td><td style='text-align: right;'>0.139198</td><td style='text-align: right;'>0.047040</td><td style='text-align: right;'>0.484039</td><td style='text-align: right;'>0.120939</td><td style='text-align: right;'>0.101934</td><td style='text-align: right;'>0.027166</td><td style='text-align: right;'>0.024252</td></tr><tr><td style='text-align: right;'>0.000143</td><td style='text-align: right;'>0.039308</td><td style='text-align: right;'>0.023379</td><td style='text-align: right;'>0.103785</td><td style='text-align: right;'>0.045175</td><td style='text-align: right;'>0.435138</td><td style='text-align: right;'>0.120628</td><td style='text-align: right;'>0.095568</td><td style='text-align: right;'>0.044587</td><td style='text-align: right;'>0.092290</td></tr></table><p>1776 cells x 10 lineages</p></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_macrostates()
g.plot_macrostates(discrete=True)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_30_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_30_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_30_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_30_1.png" />
</div>
</div>
<p>When calculating 10 macrostates we notice that 2, 7 and 8 have clear macrostates with many around them.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_macrostates(same_plot=False, ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_32_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_32_0.png" />
</div>
</div>
<p>We will now define the macrostates 2, 7 and 8 as our terminal states.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.set_terminal_states_from_macrostates([&quot;2&quot;, &quot;7&quot;, &quot;8&quot;])
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_terminal_states()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_35_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_35_0.png" />
</div>
</div>
</div>
<div class="section" id="Calculating-absorption-probabilities">
<h2>Calculating absorption probabilities<a class="headerlink" href="#Calculating-absorption-probabilities" title="Permalink to this headline">¶</a></h2>
<p>As a next step we will calculate the <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/api/cellrank.tl.estimators.GPCCA.compute_absorption_probabilities.html#cellrank.tl.estimators.GPCCA.compute_absorption_probabilities">absorption probabilities</a>.</p>
<p>For each visit, this computes the probability of being absorbed in any of the terminal_states. In particular, this corresponds to the probability that a random walk initialized in transient visit will reach any visit from a fixed transient state before reaching a visit from any other transient state.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_absorption_probabilities()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Unable to import petsc4py. For installation, please refer to: https://petsc4py.readthedocs.io/en/stable/install.html.
Defaulting to `&#39;gmres&#39;` solver.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "c752cb9dda8d479480e4138a623038e3", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_absorption_probabilities()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_39_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_39_0.png" />
</div>
</div>
<p>We can also calculate a pseudotime which is required for all patient visits to reach these states.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_absorption_probabilities(time_to_absorption=&quot;all&quot;)
g.absorption_times
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "6eb46a8ef30441b0b4911bf42cb421b1", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8b154314dcdf4237a7cddd4257df1398", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>2, 7, 8 mean</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>486.971741</td>
    </tr>
    <tr>
      <th>2</th>
      <td>545.966125</td>
    </tr>
    <tr>
      <th>3</th>
      <td>256.969513</td>
    </tr>
    <tr>
      <th>4</th>
      <td>556.640320</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>1771</th>
      <td>62.830074</td>
    </tr>
    <tr>
      <th>1772</th>
      <td>105.589584</td>
    </tr>
    <tr>
      <th>1773</th>
      <td>88.930733</td>
    </tr>
    <tr>
      <th>1774</th>
      <td>136.205994</td>
    </tr>
    <tr>
      <th>1775</th>
      <td>685.516052</td>
    </tr>
  </tbody>
</table>
<p>1776 rows × 1 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&quot;mean_time_to_absorption&quot;] = g.absorption_times[&quot;2, 7, 8 mean&quot;]
scv.pl.scatter(adata, color=&quot;mean_time_to_absorption&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_42_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_42_0.png" />
</div>
</div>
<p>As a next step we want to determine the major features driving these transitions and terminal states.</p>
<div class="section" id="Extracting-lineage-drivers">
<h3>Extracting lineage drivers<a class="headerlink" href="#Extracting-lineage-drivers" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_lineage_drivers()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>2_corr</th>
      <th>2_pval</th>
      <th>2_qval</th>
      <th>2_ci_low</th>
      <th>2_ci_high</th>
      <th>7_corr</th>
      <th>7_pval</th>
      <th>7_qval</th>
      <th>7_ci_low</th>
      <th>7_ci_high</th>
      <th>8_corr</th>
      <th>8_pval</th>
      <th>8_qval</th>
      <th>8_ci_low</th>
      <th>8_ci_high</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>age</th>
      <td>0.201495</td>
      <td>7.830093e-18</td>
      <td>NaN</td>
      <td>0.156448</td>
      <td>0.245706</td>
      <td>0.089295</td>
      <td>1.632674e-04</td>
      <td>NaN</td>
      <td>0.042959</td>
      <td>0.135246</td>
      <td>-0.236116</td>
      <td>3.929323e-24</td>
      <td>NaN</td>
      <td>-0.279560</td>
      <td>-0.191708</td>
    </tr>
    <tr>
      <th>day_28_flg</th>
      <td>0.176612</td>
      <td>5.673255e-14</td>
      <td>NaN</td>
      <td>0.131176</td>
      <td>0.221308</td>
      <td>0.026090</td>
      <td>2.718406e-01</td>
      <td>NaN</td>
      <td>-0.020448</td>
      <td>0.072516</td>
      <td>-0.172797</td>
      <td>1.986522e-13</td>
      <td>NaN</td>
      <td>-0.217562</td>
      <td>-0.127306</td>
    </tr>
    <tr>
      <th>bun_first</th>
      <td>0.159913</td>
      <td>1.107851e-11</td>
      <td>NaN</td>
      <td>0.114249</td>
      <td>0.204902</td>
      <td>-0.017014</td>
      <td>4.736988e-01</td>
      <td>NaN</td>
      <td>-0.063477</td>
      <td>0.029523</td>
      <td>-0.129899</td>
      <td>3.782014e-08</td>
      <td>NaN</td>
      <td>-0.175353</td>
      <td>-0.083892</td>
    </tr>
    <tr>
      <th>hosp_exp_flg</th>
      <td>0.155113</td>
      <td>4.564825e-11</td>
      <td>NaN</td>
      <td>0.109388</td>
      <td>0.200182</td>
      <td>0.026469</td>
      <td>2.649334e-01</td>
      <td>NaN</td>
      <td>-0.020069</td>
      <td>0.072893</td>
      <td>-0.154085</td>
      <td>6.144915e-11</td>
      <td>NaN</td>
      <td>-0.199172</td>
      <td>-0.108348</td>
    </tr>
    <tr>
      <th>chf_flg</th>
      <td>0.154022</td>
      <td>6.257784e-11</td>
      <td>NaN</td>
      <td>0.108284</td>
      <td>0.199110</td>
      <td>0.036354</td>
      <td>1.256644e-01</td>
      <td>NaN</td>
      <td>-0.010177</td>
      <td>0.082728</td>
      <td>-0.159622</td>
      <td>1.208580e-11</td>
      <td>NaN</td>
      <td>-0.204617</td>
      <td>-0.113955</td>
    </tr>
    <tr>
      <th>afib_flg</th>
      <td>0.119667</td>
      <td>4.125728e-07</td>
      <td>NaN</td>
      <td>0.073562</td>
      <td>0.165260</td>
      <td>0.059477</td>
      <td>1.216303e-02</td>
      <td>NaN</td>
      <td>0.013000</td>
      <td>0.105699</td>
      <td>-0.144443</td>
      <td>9.073144e-10</td>
      <td>NaN</td>
      <td>-0.189683</td>
      <td>-0.098592</td>
    </tr>
    <tr>
      <th>icu_exp_flg</th>
      <td>0.119113</td>
      <td>4.669401e-07</td>
      <td>NaN</td>
      <td>0.073004</td>
      <td>0.164714</td>
      <td>0.048521</td>
      <td>4.088555e-02</td>
      <td>NaN</td>
      <td>0.002012</td>
      <td>0.094821</td>
      <td>-0.136771</td>
      <td>6.818177e-09</td>
      <td>NaN</td>
      <td>-0.182126</td>
      <td>-0.090835</td>
    </tr>
    <tr>
      <th>copd_flg</th>
      <td>0.118952</td>
      <td>4.840652e-07</td>
      <td>NaN</td>
      <td>0.072841</td>
      <td>0.164555</td>
      <td>0.000402</td>
      <td>9.865041e-01</td>
      <td>NaN</td>
      <td>-0.046113</td>
      <td>0.046914</td>
      <td>-0.105168</td>
      <td>8.798814e-06</td>
      <td>NaN</td>
      <td>-0.150944</td>
      <td>-0.058943</td>
    </tr>
    <tr>
      <th>hour_icu_intime</th>
      <td>0.118112</td>
      <td>5.832398e-07</td>
      <td>NaN</td>
      <td>0.071994</td>
      <td>0.163727</td>
      <td>0.191813</td>
      <td>2.886404e-16</td>
      <td>NaN</td>
      <td>0.146608</td>
      <td>0.236219</td>
      <td>-0.229676</td>
      <td>7.087714e-23</td>
      <td>NaN</td>
      <td>-0.273271</td>
      <td>-0.185141</td>
    </tr>
    <tr>
      <th>tco2_first</th>
      <td>0.116559</td>
      <td>8.207976e-07</td>
      <td>NaN</td>
      <td>0.070427</td>
      <td>0.162193</td>
      <td>0.028174</td>
      <td>2.353696e-01</td>
      <td>NaN</td>
      <td>-0.018364</td>
      <td>0.074590</td>
      <td>-0.121231</td>
      <td>2.898899e-07</td>
      <td>NaN</td>
      <td>-0.166804</td>
      <td>-0.075141</td>
    </tr>
    <tr>
      <th>resp_flg</th>
      <td>0.105448</td>
      <td>8.324105e-06</td>
      <td>NaN</td>
      <td>0.059225</td>
      <td>0.151220</td>
      <td>-0.005363</td>
      <td>8.213449e-01</td>
      <td>NaN</td>
      <td>-0.051864</td>
      <td>0.041161</td>
      <td>-0.089489</td>
      <td>1.579417e-04</td>
      <td>NaN</td>
      <td>-0.135439</td>
      <td>-0.043155</td>
    </tr>
    <tr>
      <th>creatinine_first</th>
      <td>0.088732</td>
      <td>1.796195e-04</td>
      <td>NaN</td>
      <td>0.042393</td>
      <td>0.134689</td>
      <td>-0.002266</td>
      <td>9.239993e-01</td>
      <td>NaN</td>
      <td>-0.048774</td>
      <td>0.044253</td>
      <td>-0.076777</td>
      <td>1.198636e-03</td>
      <td>NaN</td>
      <td>-0.122851</td>
      <td>-0.030371</td>
    </tr>
    <tr>
      <th>potassium_first</th>
      <td>0.083625</td>
      <td>4.164411e-04</td>
      <td>NaN</td>
      <td>0.037256</td>
      <td>0.129634</td>
      <td>0.010143</td>
      <td>6.692928e-01</td>
      <td>NaN</td>
      <td>-0.036387</td>
      <td>0.056630</td>
      <td>-0.080394</td>
      <td>6.926764e-04</td>
      <td>NaN</td>
      <td>-0.126435</td>
      <td>-0.034007</td>
    </tr>
    <tr>
      <th>pco2_first</th>
      <td>0.077685</td>
      <td>1.046583e-03</td>
      <td>NaN</td>
      <td>0.031285</td>
      <td>0.123752</td>
      <td>-0.001770</td>
      <td>9.405766e-01</td>
      <td>NaN</td>
      <td>-0.048280</td>
      <td>0.044747</td>
      <td>-0.067358</td>
      <td>4.503565e-03</td>
      <td>NaN</td>
      <td>-0.113516</td>
      <td>-0.020910</td>
    </tr>
    <tr>
      <th>stroke_flg</th>
      <td>0.076698</td>
      <td>1.212725e-03</td>
      <td>NaN</td>
      <td>0.030292</td>
      <td>0.122773</td>
      <td>0.110363</td>
      <td>3.070279e-06</td>
      <td>NaN</td>
      <td>0.064179</td>
      <td>0.156075</td>
      <td>-0.139838</td>
      <td>3.084939e-09</td>
      <td>NaN</td>
      <td>-0.185148</td>
      <td>-0.093936</td>
    </tr>
    <tr>
      <th>renal_flg</th>
      <td>0.060937</td>
      <td>1.019778e-02</td>
      <td>NaN</td>
      <td>0.014464</td>
      <td>0.107147</td>
      <td>0.001648</td>
      <td>9.446768e-01</td>
      <td>NaN</td>
      <td>-0.044869</td>
      <td>0.048158</td>
      <td>-0.054817</td>
      <td>2.086022e-02</td>
      <td>NaN</td>
      <td>-0.101073</td>
      <td>-0.008325</td>
    </tr>
    <tr>
      <th>mal_flg</th>
      <td>0.032713</td>
      <td>1.682242e-01</td>
      <td>NaN</td>
      <td>-0.013822</td>
      <td>0.079106</td>
      <td>0.025198</td>
      <td>2.885905e-01</td>
      <td>NaN</td>
      <td>-0.021341</td>
      <td>0.071627</td>
      <td>-0.045329</td>
      <td>5.613697e-02</td>
      <td>NaN</td>
      <td>-0.091649</td>
      <td>0.001187</td>
    </tr>
    <tr>
      <th>platelet_first</th>
      <td>0.030886</td>
      <td>1.932848e-01</td>
      <td>NaN</td>
      <td>-0.015650</td>
      <td>0.077288</td>
      <td>0.022798</td>
      <td>3.369855e-01</td>
      <td>NaN</td>
      <td>-0.023740</td>
      <td>0.069239</td>
      <td>-0.042159</td>
      <td>7.569235e-02</td>
      <td>NaN</td>
      <td>-0.088499</td>
      <td>0.004363</td>
    </tr>
    <tr>
      <th>cad_flg</th>
      <td>0.030062</td>
      <td>2.054389e-01</td>
      <td>NaN</td>
      <td>-0.016475</td>
      <td>0.076469</td>
      <td>0.037896</td>
      <td>1.103913e-01</td>
      <td>NaN</td>
      <td>-0.008633</td>
      <td>0.084261</td>
      <td>-0.051303</td>
      <td>3.060805e-02</td>
      <td>NaN</td>
      <td>-0.097584</td>
      <td>-0.004801</td>
    </tr>
    <tr>
      <th>temp_1st</th>
      <td>0.023292</td>
      <td>3.266224e-01</td>
      <td>NaN</td>
      <td>-0.023247</td>
      <td>0.069730</td>
      <td>-0.039590</td>
      <td>9.533613e-02</td>
      <td>NaN</td>
      <td>-0.085946</td>
      <td>0.006936</td>
      <td>0.005347</td>
      <td>8.218680e-01</td>
      <td>NaN</td>
      <td>-0.041177</td>
      <td>0.051848</td>
    </tr>
    <tr>
      <th>bmi</th>
      <td>0.019680</td>
      <td>4.072276e-01</td>
      <td>NaN</td>
      <td>-0.026858</td>
      <td>0.066133</td>
      <td>-0.033480</td>
      <td>1.584618e-01</td>
      <td>NaN</td>
      <td>-0.079869</td>
      <td>0.013054</td>
      <td>0.004538</td>
      <td>8.484613e-01</td>
      <td>NaN</td>
      <td>-0.041985</td>
      <td>0.051041</td>
    </tr>
    <tr>
      <th>icu_los_day</th>
      <td>0.016769</td>
      <td>4.800918e-01</td>
      <td>NaN</td>
      <td>-0.029768</td>
      <td>0.063233</td>
      <td>0.002197</td>
      <td>9.262875e-01</td>
      <td>NaN</td>
      <td>-0.044321</td>
      <td>0.048706</td>
      <td>-0.016223</td>
      <td>4.945080e-01</td>
      <td>NaN</td>
      <td>-0.062689</td>
      <td>0.030314</td>
    </tr>
    <tr>
      <th>ehrapycat_day_icu_intime</th>
      <td>0.015925</td>
      <td>5.024804e-01</td>
      <td>NaN</td>
      <td>-0.030612</td>
      <td>0.062392</td>
      <td>0.016262</td>
      <td>4.934562e-01</td>
      <td>NaN</td>
      <td>-0.030274</td>
      <td>0.062729</td>
      <td>-0.024685</td>
      <td>2.985190e-01</td>
      <td>NaN</td>
      <td>-0.071117</td>
      <td>0.021854</td>
    </tr>
    <tr>
      <th>sapsi_first</th>
      <td>0.006383</td>
      <td>7.881128e-01</td>
      <td>NaN</td>
      <td>-0.040143</td>
      <td>0.052881</td>
      <td>0.034883</td>
      <td>1.417228e-01</td>
      <td>NaN</td>
      <td>-0.011650</td>
      <td>0.081265</td>
      <td>-0.028436</td>
      <td>2.310465e-01</td>
      <td>NaN</td>
      <td>-0.074850</td>
      <td>0.018102</td>
    </tr>
    <tr>
      <th>day_icu_intime_num</th>
      <td>-0.000436</td>
      <td>9.853412e-01</td>
      <td>NaN</td>
      <td>-0.046949</td>
      <td>0.046078</td>
      <td>-0.010194</td>
      <td>6.677266e-01</td>
      <td>NaN</td>
      <td>-0.056681</td>
      <td>0.036336</td>
      <td>0.007054</td>
      <td>7.664559e-01</td>
      <td>NaN</td>
      <td>-0.039473</td>
      <td>0.053550</td>
    </tr>
    <tr>
      <th>hospital_los_day</th>
      <td>-0.002376</td>
      <td>9.203218e-01</td>
      <td>NaN</td>
      <td>-0.048884</td>
      <td>0.044143</td>
      <td>-0.038168</td>
      <td>1.078489e-01</td>
      <td>NaN</td>
      <td>-0.084532</td>
      <td>0.008360</td>
      <td>0.027064</td>
      <td>2.543381e-01</td>
      <td>NaN</td>
      <td>-0.019474</td>
      <td>0.073485</td>
    </tr>
    <tr>
      <th>map_1st</th>
      <td>-0.003293</td>
      <td>8.897303e-01</td>
      <td>NaN</td>
      <td>-0.049799</td>
      <td>0.043228</td>
      <td>0.017242</td>
      <td>4.677924e-01</td>
      <td>NaN</td>
      <td>-0.029295</td>
      <td>0.063704</td>
      <td>-0.008376</td>
      <td>7.243264e-01</td>
      <td>NaN</td>
      <td>-0.054868</td>
      <td>0.038153</td>
    </tr>
    <tr>
      <th>sodium_first</th>
      <td>-0.005510</td>
      <td>8.165289e-01</td>
      <td>NaN</td>
      <td>-0.052010</td>
      <td>0.041014</td>
      <td>0.004074</td>
      <td>8.637820e-01</td>
      <td>NaN</td>
      <td>-0.042447</td>
      <td>0.050578</td>
      <td>0.002186</td>
      <td>9.266572e-01</td>
      <td>NaN</td>
      <td>-0.044332</td>
      <td>0.048695</td>
    </tr>
    <tr>
      <th>sofa_first</th>
      <td>-0.006021</td>
      <td>7.998735e-01</td>
      <td>NaN</td>
      <td>-0.052519</td>
      <td>0.040504</td>
      <td>-0.054387</td>
      <td>2.188514e-02</td>
      <td>NaN</td>
      <td>-0.100646</td>
      <td>-0.007894</td>
      <td>0.040899</td>
      <td>8.486949e-02</td>
      <td>NaN</td>
      <td>-0.005625</td>
      <td>0.087247</td>
    </tr>
    <tr>
      <th>weight_first</th>
      <td>-0.015079</td>
      <td>5.254450e-01</td>
      <td>NaN</td>
      <td>-0.061549</td>
      <td>0.031457</td>
      <td>-0.053703</td>
      <td>2.360785e-02</td>
      <td>NaN</td>
      <td>-0.099967</td>
      <td>-0.007207</td>
      <td>0.048431</td>
      <td>4.126220e-02</td>
      <td>NaN</td>
      <td>0.001922</td>
      <td>0.094731</td>
    </tr>
    <tr>
      <th>abg_count</th>
      <td>-0.018195</td>
      <td>4.435428e-01</td>
      <td>NaN</td>
      <td>-0.064654</td>
      <td>0.028343</td>
      <td>-0.035655</td>
      <td>1.331036e-01</td>
      <td>NaN</td>
      <td>-0.082033</td>
      <td>0.010876</td>
      <td>0.039376</td>
      <td>9.714595e-02</td>
      <td>NaN</td>
      <td>-0.007151</td>
      <td>0.085732</td>
    </tr>
    <tr>
      <th>liver_flg</th>
      <td>-0.036613</td>
      <td>1.229835e-01</td>
      <td>NaN</td>
      <td>-0.082986</td>
      <td>0.009917</td>
      <td>-0.036820</td>
      <td>1.208872e-01</td>
      <td>NaN</td>
      <td>-0.083191</td>
      <td>0.009711</td>
      <td>0.056389</td>
      <td>1.745968e-02</td>
      <td>NaN</td>
      <td>0.009901</td>
      <td>0.102633</td>
    </tr>
    <tr>
      <th>hgb_first</th>
      <td>-0.042999</td>
      <td>7.003851e-02</td>
      <td>NaN</td>
      <td>-0.089334</td>
      <td>0.003522</td>
      <td>-0.026249</td>
      <td>2.689299e-01</td>
      <td>NaN</td>
      <td>-0.072674</td>
      <td>0.020289</td>
      <td>0.055091</td>
      <td>2.022858e-02</td>
      <td>NaN</td>
      <td>0.008600</td>
      <td>0.101345</td>
    </tr>
    <tr>
      <th>hr_1st</th>
      <td>-0.053644</td>
      <td>2.376197e-02</td>
      <td>NaN</td>
      <td>-0.099908</td>
      <td>-0.007148</td>
      <td>-0.070570</td>
      <td>2.916336e-03</td>
      <td>NaN</td>
      <td>-0.116700</td>
      <td>-0.024135</td>
      <td>0.093481</td>
      <td>7.889879e-05</td>
      <td>NaN</td>
      <td>0.047172</td>
      <td>0.139388</td>
    </tr>
    <tr>
      <th>chloride_first</th>
      <td>-0.066060</td>
      <td>5.342364e-03</td>
      <td>NaN</td>
      <td>-0.112228</td>
      <td>-0.019606</td>
      <td>0.006772</td>
      <td>7.755263e-01</td>
      <td>NaN</td>
      <td>-0.039754</td>
      <td>0.053269</td>
      <td>0.053824</td>
      <td>2.329483e-02</td>
      <td>NaN</td>
      <td>0.007329</td>
      <td>0.100087</td>
    </tr>
    <tr>
      <th>wbc_first</th>
      <td>-0.072184</td>
      <td>2.328812e-03</td>
      <td>NaN</td>
      <td>-0.118300</td>
      <td>-0.025757</td>
      <td>-0.022526</td>
      <td>3.427980e-01</td>
      <td>NaN</td>
      <td>-0.068967</td>
      <td>0.024013</td>
      <td>0.078402</td>
      <td>9.394864e-04</td>
      <td>NaN</td>
      <td>0.032005</td>
      <td>0.124461</td>
    </tr>
    <tr>
      <th>aline_flg</th>
      <td>-0.075074</td>
      <td>1.539596e-03</td>
      <td>NaN</td>
      <td>-0.121165</td>
      <td>-0.028661</td>
      <td>-0.041249</td>
      <td>8.223336e-02</td>
      <td>NaN</td>
      <td>-0.087595</td>
      <td>0.005274</td>
      <td>0.093203</td>
      <td>8.287724e-05</td>
      <td>NaN</td>
      <td>0.046893</td>
      <td>0.139114</td>
    </tr>
    <tr>
      <th>gender_num</th>
      <td>-0.080245</td>
      <td>7.087512e-04</td>
      <td>NaN</td>
      <td>-0.126288</td>
      <td>-0.033858</td>
      <td>-0.022742</td>
      <td>3.381749e-01</td>
      <td>NaN</td>
      <td>-0.069183</td>
      <td>0.023796</td>
      <td>0.085652</td>
      <td>2.998515e-04</td>
      <td>NaN</td>
      <td>0.039295</td>
      <td>0.131642</td>
    </tr>
    <tr>
      <th>spo2_1st</th>
      <td>-0.080428</td>
      <td>6.890013e-04</td>
      <td>NaN</td>
      <td>-0.126469</td>
      <td>-0.034042</td>
      <td>0.007407</td>
      <td>7.551305e-01</td>
      <td>NaN</td>
      <td>-0.039120</td>
      <td>0.053902</td>
      <td>0.066088</td>
      <td>5.322884e-03</td>
      <td>NaN</td>
      <td>0.019634</td>
      <td>0.112256</td>
    </tr>
    <tr>
      <th>po2_first</th>
      <td>-0.090859</td>
      <td>1.248525e-04</td>
      <td>NaN</td>
      <td>-0.136795</td>
      <td>-0.044534</td>
      <td>0.012101</td>
      <td>6.103514e-01</td>
      <td>NaN</td>
      <td>-0.034432</td>
      <td>0.058582</td>
      <td>0.072219</td>
      <td>2.317203e-03</td>
      <td>NaN</td>
      <td>0.025792</td>
      <td>0.118335</td>
    </tr>
    <tr>
      <th>ehrapycat_service_unit</th>
      <td>-0.101271</td>
      <td>1.878389e-05</td>
      <td>NaN</td>
      <td>-0.147092</td>
      <td>-0.055016</td>
      <td>-0.004370</td>
      <td>8.540039e-01</td>
      <td>NaN</td>
      <td>-0.050873</td>
      <td>0.042152</td>
      <td>0.092180</td>
      <td>9.923161e-05</td>
      <td>NaN</td>
      <td>0.045863</td>
      <td>0.138101</td>
    </tr>
    <tr>
      <th>service_num</th>
      <td>-0.122847</td>
      <td>2.004048e-07</td>
      <td>NaN</td>
      <td>-0.168398</td>
      <td>-0.076772</td>
      <td>-0.003277</td>
      <td>8.902491e-01</td>
      <td>NaN</td>
      <td>-0.049783</td>
      <td>0.043243</td>
      <td>0.110493</td>
      <td>2.988408e-06</td>
      <td>NaN</td>
      <td>0.064310</td>
      <td>0.156204</td>
    </tr>
    <tr>
      <th>mort_day_censored</th>
      <td>-0.169885</td>
      <td>5.070734e-13</td>
      <td>NaN</td>
      <td>-0.214702</td>
      <td>-0.124354</td>
      <td>-0.052856</td>
      <td>2.590247e-02</td>
      <td>NaN</td>
      <td>-0.099126</td>
      <td>-0.006358</td>
      <td>0.184375</td>
      <td>4.053053e-15</td>
      <td>NaN</td>
      <td>0.139054</td>
      <td>0.228925</td>
    </tr>
    <tr>
      <th>censor_flg</th>
      <td>-0.192525</td>
      <td>2.228833e-16</td>
      <td>NaN</td>
      <td>-0.236917</td>
      <td>-0.147330</td>
      <td>-0.055917</td>
      <td>1.842558e-02</td>
      <td>NaN</td>
      <td>-0.102165</td>
      <td>-0.009428</td>
      <td>0.206339</td>
      <td>1.198193e-18</td>
      <td>NaN</td>
      <td>0.161374</td>
      <td>0.250449</td>
    </tr>
    <tr>
      <th>iv_day_1</th>
      <td>-0.661226</td>
      <td>1.129444e-245</td>
      <td>NaN</td>
      <td>-0.686622</td>
      <td>-0.634218</td>
      <td>-0.319777</td>
      <td>2.965727e-44</td>
      <td>NaN</td>
      <td>-0.360922</td>
      <td>-0.277389</td>
      <td>0.792408</td>
      <td>0.000000e+00</td>
      <td>NaN</td>
      <td>0.774438</td>
      <td>0.809100</td>
    </tr>
    <tr>
      <th>sepsis_flg</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_lineage_drivers(lineage=&quot;2&quot;, ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_46_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_46_0.png" />
</div>
</div>
<p>Cluster 2 seems to have a lot of patients with COPD, congestive heart failure (chf) and atrial fibrillation (afib) as a comorbidities.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_lineage_drivers(lineage=&quot;7&quot;, ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_48_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_48_0.png" />
</div>
</div>
<p>Cluster 7 is primarily influenced by atrial fibrillation (afib), coronary artery disease (cad), congestive heart failure (chf) and the SAPS I score (sapsi).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_lineage_drivers(lineage=&quot;8&quot;, ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_50_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_50_0.png" />
</div>
</div>
<p>Cluster 8 is primarly defined as a terminal state by the heart rate (hr) and the IV input fluids (iv).</p>
<p>Although some of the features (e.g. congestive heart failure and age) overlap as drivers for death as a terminal state, all three clusters have unique origins of their macrostates.</p>
<p>We can also easily examine how likely it is that patients from a specific cluster end up in one of our three death dominated clusters.</p>
</div>
</div>
<div class="section" id="Exploring-cluster-fates">
<h2>Exploring cluster fates<a class="headerlink" href="#Exploring-cluster-fates" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.cluster_fates(adata, mode=&quot;bar&quot;, cluster_key=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_55_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_55_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.cluster_fates(adata, mode=&quot;heatmap&quot;, cluster_key=&quot;leiden_0_5&quot;)
cr.pl.cluster_fates(adata, mode=&quot;clustermap&quot;, cluster_key=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_56_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_56_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_56_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_56_1.png" />
</div>
</div>
<p>We can also leverage PAGA to visualize this information in a PAGA graph.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.tl.paga(adata, groups=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.cluster_fates(adata, mode=&quot;paga_pie&quot;, basis=&quot;umap&quot;, cluster_key=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: transitions_confidence not found, using connectivites instead.
WARNING: Invalid color key. Using grey instead.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/networkx/drawing/nx_pylab.py:717: MatplotlibDeprecationWarning: Passing *transOffset* without *offsets* has no effect. This behavior is deprecated since 3.5 and in 3.6, *transOffset* will begin having an effect regardless of *offsets*. In the meantime, if you wish to set *transOffset*, call collection.set_offset_transform(transOffset) explicitly.
  edge_collection = mpl.collections.LineCollection(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_59_2.png" src="../../_images/tutorials_notebooks_mimic_2_fate_59_2.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.cluster_fates(adata, mode=&quot;paga&quot;, legend_loc=&quot;on data&quot;, basis=&quot;umap&quot;, cluster_key=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: transitions_confidence not found, using connectivites instead.
WARNING: Invalid color key. Using grey instead.
WARNING: transitions_confidence not found, using connectivites instead.
WARNING: Invalid color key. Using grey instead.
WARNING: transitions_confidence not found, using connectivites instead.
WARNING: Invalid color key. Using grey instead.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/networkx/drawing/nx_pylab.py:717: MatplotlibDeprecationWarning: Passing *transOffset* without *offsets* has no effect. This behavior is deprecated since 3.5 and in 3.6, *transOffset* will begin having an effect regardless of *offsets*. In the meantime, if you wish to set *transOffset*, call collection.set_offset_transform(transOffset) explicitly.
  edge_collection = mpl.collections.LineCollection(
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/networkx/drawing/nx_pylab.py:717: MatplotlibDeprecationWarning: Passing *transOffset* without *offsets* has no effect. This behavior is deprecated since 3.5 and in 3.6, *transOffset* will begin having an effect regardless of *offsets*. In the meantime, if you wish to set *transOffset*, call collection.set_offset_transform(transOffset) explicitly.
  edge_collection = mpl.collections.LineCollection(
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/networkx/drawing/nx_pylab.py:717: MatplotlibDeprecationWarning: Passing *transOffset* without *offsets* has no effect. This behavior is deprecated since 3.5 and in 3.6, *transOffset* will begin having an effect regardless of *offsets*. In the meantime, if you wish to set *transOffset*, call collection.set_offset_transform(transOffset) explicitly.
  edge_collection = mpl.collections.LineCollection(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_60_2.png" src="../../_images/tutorials_notebooks_mimic_2_fate_60_2.png" />
</div>
</div>
<p>Finally, we can use a <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/auto_examples/plotting/plot_circular_embedding.html">circular projection</a> to represent the same information.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.circular_projection(
    adata, keys=[&quot;leiden_0_5&quot;, &quot;kl_divergence&quot;], legend_loc=&quot;upper right&quot;
)
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_62_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_62_0.png" />
</div>
</div>
</div>
</div>
<div class="section" id="Determining-patient-fate-with-a-PseudotimeKernel">
<h1>Determining patient fate with a PseudotimeKernel<a class="headerlink" href="#Determining-patient-fate-with-a-PseudotimeKernel" title="Permalink to this headline">¶</a></h1>
<p>We have now learnt that for example death dominated cluster 2 origins from cluster 0. Hence, we will now set a patient from cluster 0 as our root for our pseudotime calculation and will verify that we would indeed end up in cluster 2.</p>
<p>Using pseudotime to find terminal states is even more useful when the terminal states are unknown and only a root cluster is clear.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = ep.dt.mimic_2(encoded=True)
ep.pp.knn_impute(adata)
ep.pp.pca(adata)
ep.pp.neighbors(adata)
ep.tl.umap(adata)
ep.tl.leiden(adata, resolution=0.5, key_added=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pl.umap(adata, color=[&quot;leiden_0_5&quot;], title=&quot;Leiden 0.5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;service_unit&#39; as categorical
... storing &#39;day_icu_intime&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_66_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_66_1.png" />
</div>
</div>
<div class="section" id="Pseudotime-Kernels-and-stream-embeddings">
<h2>Pseudotime Kernels and stream embeddings<a class="headerlink" href="#Pseudotime-Kernels-and-stream-embeddings" title="Permalink to this headline">¶</a></h2>
<p>The Pseudotime kernel computes direct transition probabilities based on a KNN graph and pseudotime.</p>
<p>The KNN graph contains information about the (undirected) connectivities among cells, reflecting their similarity. Pseudotime can be used to either remove edges that point against the direction of increasing pseudotime, or to downweight them.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.uns[&#39;iroot&#39;] = np.flatnonzero(adata.obs[&#39;leiden_0_5&#39;] == &#39;0&#39;)[0]
ep.tl.dpt(adata)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pk = PseudotimeKernel(adata)
pk.compute_transition_matrix()
pk.compute_projection(basis=&quot;umap&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "42bdf8440d0145fc836054e7a82fc7b4", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scv.pl.velocity_embedding_stream(adata, vkey=&quot;T_fwd&quot;, basis=&quot;umap&quot;, color=&quot;leiden_0_5&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_71_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_71_0.png" />
</div>
</div>
<p>Now the arrows are much more clear and point into the direction of cluster 2 from cluster 0. Apparently transitions from cluster 11 to cluster 7 are possible, but also transitions via clusters 1, 6, 4, 3 and 8.</p>
</div>
<div class="section" id="Simulating-transitions-with-random-walks">
<h2>Simulating transitions with random walks<a class="headerlink" href="#Simulating-transitions-with-random-walks" title="Permalink to this headline">¶</a></h2>
<p>Cellrank makes it easy to simulate the behavior of random walks from specific clusters. This allows us to not only visualize where the patients end up, but also roughly how many in which clusters after a defined number of iterations. We can either just start walking…</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pk.plot_random_walks(
    100,
    start_ixs={&quot;leiden_0_5&quot;: &quot;0&quot;},
    max_iter=100,
    show_progress_bar=False,
    ixs_legend_loc=&quot;best&quot;,
    seed=42,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_75_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_75_0.png" />
</div>
</div>
<p>… or set a number of required hits in one or more terminal clusters. Here, we require 50 hits in one of our three “death” clusters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pk.plot_random_walks(
    200,
    start_ixs={&quot;leiden_0_5&quot;: &quot;0&quot;},
    stop_ixs={&quot;leiden_0_5&quot;: [&quot;8&quot;, &quot;7&quot;, &quot;2&quot;]},
    max_iter=100,
    successive_hits=50,
    show_progress_bar=False,
    cmap=&quot;viridis&quot;,
    seed=42,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_77_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_77_0.png" />
</div>
</div>
</div>
<div class="section" id="id1">
<h2>Determining macrostates and terminal states<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>Similarly to above when using our ConnectivityKernel we can compute macrostate memberships.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g = cr.tl.estimators.GPCCA(pk)
g.compute_schur()
g.compute_macrostates(cluster_key=&quot;leiden_0_5&quot;)
g.macrostates_memberships
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Unable to import `petsc4py` or `slepc4py`. Using `method=&#39;brandts&#39;`
WARNING: For `method=&#39;brandts&#39;`, dense matrix is required. Densifying
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div style='scoped' class='rendered_html'><table class='dataframe'><tr><th style='background-color: #279e68; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px'>2</th><th style='background-color: #ffbb78; color: #000000; text-align: center; word-wrap: break-word; max-width: 100px'>10</th></tr><tr><td style='text-align: right;'>0.000000</td><td style='text-align: right;'>1.000000</td></tr><tr><td style='text-align: right;'>0.010048</td><td style='text-align: right;'>0.989952</td></tr><tr><td style='text-align: right;'>0.000000</td><td style='text-align: right;'>1.000000</td></tr><tr><td style='text-align: right;'>0.000000</td><td style='text-align: right;'>1.000000</td></tr><tr><td style='text-align: right;'>0.000000</td><td style='text-align: right;'>1.000000</td></tr><tr><td style='text-align: right;'>0.000001</td><td style='text-align: right;'>0.999999</td></tr><tr><td style='text-align: right;'>0.012068</td><td style='text-align: right;'>0.987932</td></tr><tr><td style='text-align: right;'>0.000319</td><td style='text-align: right;'>0.999681</td></tr><tr><td style='text-align: right;'>0.000856</td><td style='text-align: right;'>0.999144</td></tr><tr><td style='text-align: right;'>0.000000</td><td style='text-align: right;'>1.000000</td></tr><tr><td style='text-align: right;'>...</td><td style='text-align: right;'>...</td></tr><tr><td style='text-align: right;'>0.000000</td><td style='text-align: right;'>1.000000</td></tr><tr><td style='text-align: right;'>0.171079</td><td style='text-align: right;'>0.828921</td></tr><tr><td style='text-align: right;'>0.041941</td><td style='text-align: right;'>0.958059</td></tr><tr><td style='text-align: right;'>0.000000</td><td style='text-align: right;'>1.000000</td></tr><tr><td style='text-align: right;'>0.871996</td><td style='text-align: right;'>0.128004</td></tr><tr><td style='text-align: right;'>0.000000</td><td style='text-align: right;'>1.000000</td></tr><tr><td style='text-align: right;'>0.937809</td><td style='text-align: right;'>0.062191</td></tr><tr><td style='text-align: right;'>0.809891</td><td style='text-align: right;'>0.190109</td></tr><tr><td style='text-align: right;'>1.000000</td><td style='text-align: right;'>0.000000</td></tr><tr><td style='text-align: right;'>0.893367</td><td style='text-align: right;'>0.106633</td></tr></table><p>1776 cells x 2 lineages</p></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_macrostates()
g.plot_macrostates(discrete=True)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_81_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_81_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_81_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_81_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[75]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_macrostates(same_plot=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_82_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_82_0.png" />
</div>
</div>
<p>As expected cluster 2 is a macrostate. Furthermore, cluster 10 is apparently a macrostates. This can likely be attributed to the fact that the people in cluster 10 are IV+.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[76]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pl.umap(adata, color=&quot;iv_day_1&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_84_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_84_0.png" />
</div>
</div>
</div>
<div class="section" id="id2">
<h2>Calculating absorption probabilities<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<p>Analogously we can compute terminal states, the absorption probabilities and the absorption times.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[77]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_terminal_states()
g.plot_terminal_states()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_87_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_87_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[78]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_absorption_probabilities(preconditioner=&#39;ilu&#39;, tol=1e-10)
g.plot_absorption_probabilities()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "9e877e3e017b4d00b470ed121ece068b", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_88_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_88_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_absorption_probabilities(time_to_absorption=&quot;all&quot;, preconditioner=&#39;ilu&#39;, tol=1e-10)
g.absorption_times
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "334351a674ee432d8258ec1144fccb06", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d48ed977be8b428bbd20012f44d538c7", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0_1, 0_2, 2, 6, 10 mean</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>179.287567</td>
    </tr>
    <tr>
      <th>1</th>
      <td>112.491760</td>
    </tr>
    <tr>
      <th>2</th>
      <td>286.910187</td>
    </tr>
    <tr>
      <th>3</th>
      <td>146.307037</td>
    </tr>
    <tr>
      <th>4</th>
      <td>262.389893</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>1771</th>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1772</th>
      <td>116.465927</td>
    </tr>
    <tr>
      <th>1773</th>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1774</th>
      <td>107.240761</td>
    </tr>
    <tr>
      <th>1775</th>
      <td>87.522881</td>
    </tr>
  </tbody>
</table>
<p>1776 rows × 1 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[69]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&quot;mean_time_to_absorption&quot;] = g.absorption_times[&quot;0_1, 0_2, 2, 6, 10 mean&quot;]
scv.pl.scatter(adata, color=&quot;mean_time_to_absorption&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_90_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_90_0.png" />
</div>
</div>
</div>
<div class="section" id="id3">
<h2>Extracting lineage drivers<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h2>
<p>Furthermore, just like above we can visualize the lineage drivers.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_lineage_drivers()
g.plot_lineage_drivers(lineage=&quot;2&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_93_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_93_0.png" />
</div>
</div>
<p>The primary lineage drivers for cluster 2 are atrial fibrillation (afib), congestive heart failure (chf), blood urea nitrogen (bun) and COPD.</p>
</div>
<div class="section" id="Determining-feature-trends">
<h2>Determining feature trends<a class="headerlink" href="#Determining-feature-trends" title="Permalink to this headline">¶</a></h2>
<p>cellrank uses <a class="reference external" href="https://cellrank.readthedocs.io/en/latest/api/cellrank.ul.models.GAM.html">Generalized Additive Models (GAMs)</a> to determine trends of features.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[71]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model = cr.ul.models.GAM(adata)
cr.pl.gene_trends(
    adata,
    model,
    [&quot;iv_day_1&quot;],
    time_key=&quot;dpt_pseudotime&quot;,
    show_progress_bar=False,
)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "ca04616867704242bf04bdef4e7f9e8f", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_97_2.png" src="../../_images/tutorials_notebooks_mimic_2_fate_97_2.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[80]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.gene_trends(
    adata,
    model,
    [&quot;iv_day_1&quot;],
    same_plot=True,
    hide_cells=True,
    time_key=&quot;dpt_pseudotime&quot;,
    show_progress_bar=False,
)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "07ff68cf5c47474b8c6b1f2336c52248", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_98_2.png" src="../../_images/tutorials_notebooks_mimic_2_fate_98_2.png" />
</div>
</div>
<p>As expected, the IV value is only marginally important for macro state 2, but even more important for macrostate 10. This is reflected in the required pseudotime to reach macrostate 10 with increasing IV values.</p>
</div>
</div>


              </div>
              
              
              <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="mimic_2_introduction.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">MIMIC-II IAC Introduction</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../../contributing.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Contributor Guide</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
              
          </main>
          

      </div>
    </div>
  
    <script src="../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>
  <footer class="footer mt-5 mt-md-0">
  <div class="container">
    
    <div class="footer-item">
      <p class="copyright">
    &copy; Copyright 2021, Lukas Heumos, Theislab.<br>
</p>
    </div>
    
    <div class="footer-item">
      <p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 4.4.0.<br>
</p>
    </div>
    
  </div>
</footer>
  </body>
</html>