<!doctype html>
<html class="no-js">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

<meta property="og:title" content="MIMIC-II Patient Fate" />
  
<meta property="og:type" content="website" />
  
<meta property="og:url" content="https://ehrapy.readthedocs.io/en/latest/tutorials/notebooks/mimic_2_fate.html" />
  
<meta property="og:description" content="This tutorial focuses on the determination of patient fate. The goal is to detect terminal states and the corresponding origins based on pseudotime. Very rarely does the actual real time reflect th..." />
  
<meta property="og:image" content="https://ehrapy.readthedocs.io/en/latest//_static/logo.png" />
  
<meta property="og:image:alt" content="MIMIC-II Patient Fate" />
  <link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" /><link rel="next" title="Diabetes 130 Vignette" href="diabetes_130.html" /><link rel="prev" title="MIMIC-II IAC Introduction" href="mimic_2_introduction.html" />

    <meta name="generator" content="sphinx-4.5.0, furo 2022.04.07"/>
        <title>MIMIC-II Patient Fate - ehrapy</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?digest=68f4518137b9aefe99b631505a2064c3c42c9852" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/override.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/sphinx_gallery.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">ehrapy</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand centered" href="../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../_static/ehrapy_pure.png" alt="Logo"/>
  </div>
  
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/usage.html">Usage</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../index.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="ehrapy_introduction.html">ehrapy Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="ehrapy_introduction.html#Fundamental-Principles">Fundamental Principles</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html">MIMIC-II IAC Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Preprocessing">Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#PCA-&amp;-UMAP">PCA &amp; UMAP</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Batch-effect-check">Batch effect check</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Leiden-clustering">Leiden clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Hospital-&amp;-ICU-statistics">Hospital &amp; ICU statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Death">Death</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Demographics">Demographics</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Comorbidities">Comorbidities</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Initial-lab-measurements">Initial lab measurements</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Annotation">Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="mimic_2_introduction.html#Paga">Paga</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">MIMIC-II Patient Fate</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Determining-patient-fate-using-a-ConnectivityKernel">Determining patient fate using a ConnectivityKernel</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Determining-patient-fate-with-a-PseudotimeKernel">Determining patient fate with a PseudotimeKernel</a></li>
<li class="toctree-l2"><a class="reference internal" href="diabetes_130.html">Diabetes 130 Vignette</a></li>
<li class="toctree-l2"><a class="reference internal" href="diabetes_130.html#Visual-Exploration">Visual Exploration</a></li>
<li class="toctree-l2"><a class="reference internal" href="medcat.html">Extracting from free text with MedCAT</a></li>
<li class="toctree-l2"><a class="reference internal" href="medcat.html#Custom-MedCAT-object-of-ehrapy">Custom MedCAT object of ehrapy</a></li>
<li class="toctree-l2"><a class="reference internal" href="medcat.html#Extracting-and-visualizing-all-disease-entities">Extracting and visualizing all disease entities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributor Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../code_of_conduct.html">Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/theislab/ehrapy/discussions">Discussions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../references.html">References</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container"><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<style>
    p {
        margin-bottom: 0.5rem;
    }
    /* Main index page overview cards */
    /* https://github.com/spatialaudio/nbsphinx/pull/635/files */
    .jp-RenderedHTMLCommon table,
    div.rendered_html table {
    border: none;
    border-collapse: collapse;
    border-spacing: 0;
    font-size: 12px;
    table-layout: fixed;
    color: inherit;
    }

    body:not([data-theme=light]) .jp-RenderedHTMLCommon tbody tr:nth-child(odd),
    body:not([data-theme=light]) div.rendered_html tbody tr:nth-child(odd) {
    background: rgba(255, 255, 255, .1);
    }
</style><div class="admonition note">
<p class="admonition-title">Note</p>
<p>
    This page was generated from
    <a class="reference external" href="https://github.com/theislab/ehrapy/tree/0.2.0/">mimic_2_fate.ipynb</a>.
    Some tutorial content may look better in light mode.
    </p>
</div><section id="MIMIC-II-Patient-Fate">
<h1>MIMIC-II Patient Fate<a class="headerlink" href="#MIMIC-II-Patient-Fate" title="Permalink to this headline">#</a></h1>
<p>This tutorial focuses on the determination of patient fate. The goal is to detect terminal states and the corresponding origins based on pseudotime. Very rarely does the actual real time reflect the actual progression of disease. Imagine that 100 patients come in at day 45 and get measurements done. Some will be healthy still (or recovered), some at the onset of a specific disease and some at the height of it. What we are much more interested in is a continuous transition of states (e.g. healthy
to diseased to dead). This can be achieved by identifying source states (e.g. healthy) and then calculating pseudotime from this state. Based on markov chain modelling we uncover patient dynamics using <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/index.html">cellrank</a>. For more details please read the <a class="reference external" href="https://www.nature.com/articles/s41592-021-01346-6">Cellrank paper</a>.</p>
<p>In this tutorial we will be using cellrank to:</p>
<ol class="arabic simple">
<li><p>Compute initial &amp; terminal as well as intermediate macrostates of our patient landscape.</p></li>
<li><p>Infer fate proabilities towards the terminal states for each individual patient (visit).</p></li>
<li><p>Visualize feature trends along specific patient states while accounting for the continuous nature of fate determination.</p></li>
<li><p>Identify potential driver features for each identified trajectory.</p></li>
</ol>
<p>A skim of the extensive and well written <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/index.html">cellrank documentation</a> is recommended. Especially the <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/cellrank_basics.html">basic tutorial</a> chapter is useful. If you are not familiar with single-cell data do not be afraid and replace cells with patients and genes with features in your mind.</p>
<p>This tutorial requires <code class="docutils literal notranslate"><span class="pre">cellrank</span></code> and <code class="docutils literal notranslate"><span class="pre">scvelo</span></code> to be installed. These are not dependencies of ehrapy and must be installed separately:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install cellrank scvelo
</pre></div>
</div>
<p>This tutorial uses the MIMIC-II IAC dataset. It was created for the purpose of a case study in the book: Secondary Analysis of Electronic Health Records, published by Springer in 2016. In particular, the dataset was used throughout Chapter 16 (Data Analysis) by Raffa J. et al. to investigate the effectiveness of indwelling arterial catheters in hemodynamically stable patients with respiratory failure for mortality outcomes. The dataset is derived from MIMIC-II, the publicly-accessible critical
care database. It contains summary clinical data and outcomes for 1,776 patients.</p>
<p>More details on the dataset can be found here: <a class="reference external" href="https://physionet.org/content/mimic2-iaccd/1.0/">https://physionet.org/content/mimic2-iaccd/1.0/</a>.</p>
<p>This tutorial will eventually load a preprocessed version of the MIMIC-II dataset, but for now we will start from scratch.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!pip install cellrank scvelo
</pre></div>
</div>
</div>
<section id="Importing-ehrapy-and-setting-plotting-parameters">
<h2>Importing ehrapy and setting plotting parameters<a class="headerlink" href="#Importing-ehrapy-and-setting-plotting-parameters" title="Permalink to this headline">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import ehrapy as ep
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import cellrank as cr
from cellrank.tl.kernels import ConnectivityKernel
from cellrank.tl.kernels import PseudotimeKernel
import scvelo as scv
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">Installed version </span><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">0.2</span><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">.</span><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">0</span><span style="color: #808000; text-decoration-color: #808000; font-weight: bold"> of ehrapy is newer than the latest release </span><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">0.1</span><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">.</span><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">0</span><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">! You are running a </span>
<span style="color: #808000; text-decoration-color: #808000; font-weight: bold">nightly version and features may break!</span>
</pre></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import warnings
from matplotlib import MatplotlibDeprecationWarning

warnings.simplefilter("ignore", category=UserWarning)
warnings.simplefilter("ignore", category=FutureWarning)
warnings.simplefilter("ignore", category=DeprecationWarning)
warnings.simplefilter("ignore", category=MatplotlibDeprecationWarning)
</pre></div>
</div>
</div>
</section>
<section id="Getting-and-preprocessing-the-MIMIC-II-dataset">
<h2>Getting and preprocessing the MIMIC-II dataset<a class="headerlink" href="#Getting-and-preprocessing-the-MIMIC-II-dataset" title="Permalink to this headline">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = ep.dt.mimic_2(encoded=True)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "936560f1ddc0463aa388fda32ffc392c", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pp.knn_impute(adata)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">Quality control metrics missing. Calculating...</span>
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d7a302ce108a4cf1896b6059ef789d9d", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 1776 × 46
    obs: 'service_unit', 'day_icu_intime', 'missing_values_abs', 'missing_values_pct'
    var: 'missing_values_abs', 'missing_values_pct', 'mean', 'median', 'standard_deviation', 'min', 'max'
    uns: 'numerical_columns', 'non_numerical_columns', 'original_values_categoricals', 'var_to_encoding', 'encoding_to_var'
    layers: 'original'
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pp.pca(adata)
ep.pp.neighbors(adata)
ep.tl.umap(adata)
ep.tl.leiden(adata, resolution=0.5, key_added="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pl.umap(adata, color=["leiden_0_5"], title="Leiden 0.5")
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing 'service_unit' as categorical
... storing 'day_icu_intime' as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_12_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_12_1.png"/>
</div>
</div>
</section>
</section>
<section id="Determining-patient-fate-using-a-ConnectivityKernel">
<h1>Determining patient fate using a ConnectivityKernel<a class="headerlink" href="#Determining-patient-fate-using-a-ConnectivityKernel" title="Permalink to this headline">#</a></h1>
<section id="Connectivity-Kernels-and-stream-embeddings">
<h2>Connectivity Kernels and stream embeddings<a class="headerlink" href="#Connectivity-Kernels-and-stream-embeddings" title="Permalink to this headline">#</a></h2>
<p>Depending on the data it may not always be possible to clearly define a cluster or specific patients as the origin. When working with single-cell data this is easier because when stem cells are detected these are usually the origin of cell differentiation processes.</p>
<p>In our case we do not have any clear origin and will therefore start with <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/classes.html#connectivity-kernel">cellrank’s ConnectivityKernel</a>. This kernel computes transition probabilities based on similarities among patients using a KNN graph.</p>
<p>First we define the kernel, compute the transition matrix and a projection ontop of the UMAP.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ck = ConnectivityKernel(adata)
ck.compute_transition_matrix()
ck.compute_projection(basis="umap")
</pre></div>
</div>
</div>
<p>Note that ConnectivityKernel has a <code class="docutils literal notranslate"><span class="pre">backward</span></code> parameter which can be set to <code class="docutils literal notranslate"><span class="pre">True</span></code> to compute everything backwards in time.</p>
<p>We can now visualize the project forwards and backwards in time using <a class="reference external" href="https://scvelo.readthedocs.io/">scvelo</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scv.pl.velocity_embedding_stream(adata, vkey="T_fwd", basis="umap", color="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_21_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_21_0.png"/>
</div>
</div>
<p>We observe that several clusters including e.g. 8 only really have outgoing arrows indicating steady states.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pl.umap(adata, color="hosp_exp_flg", title="Died in Hospital")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_23_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_23_0.png"/>
</div>
</div>
<p>Cluster 8 is apparently the cluster where most people died and clusters 2 and 7 further include many deceased patients.</p>
<p>Let’s try to find the origins of the “death clusters”. We will now define a <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/api/cellrank.tl.estimators.GPCCA.html">GPCCA</a> estimator to predict the patient fates using the above calculated transition matrix.</p>
</section>
<section id="Determining-macrostates-and-terminal-states">
<h2>Determining macrostates and terminal states<a class="headerlink" href="#Determining-macrostates-and-terminal-states" title="Permalink to this headline">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g = cr.tl.estimators.GPCCA(ck)
</pre></div>
</div>
</div>
<p>As a first step we try to identify macrostates in the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_macrostates(n_states=10, cluster_key="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Unable to import `petsc4py` or `slepc4py`. Using `method='brandts'`
WARNING: For `method='brandts'`, dense matrix is required. Densifying
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.macrostates_memberships
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div class="rendered_html" style="scoped"><div class="table-wrapper"><table class="dataframe"><tr><th style="background-color: #b25454; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">10_1</th><th style="background-color: #e6c56c; color: #000000; text-align: center; word-wrap: break-word; max-width: 100px">10_2</th><th style="background-color: #17becf; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">8</th><th style="background-color: #b5bd61; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">7</th><th style="background-color: #aec7e8; color: #000000; text-align: center; word-wrap: break-word; max-width: 100px">9</th><th style="background-color: #d62728; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">3</th><th style="background-color: #279e68; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">2</th><th style="background-color: #aa40fc; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">4</th><th style="background-color: #ff7f0e; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">1</th><th style="background-color: #1f77b4; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">0</th></tr><tr><td style="text-align: right;">0.003050</td><td style="text-align: right;">0.032411</td><td style="text-align: right;">0.442545</td><td style="text-align: right;">0.106267</td><td style="text-align: right;">0.081307</td><td style="text-align: right;">0.127711</td><td style="text-align: right;">0.089226</td><td style="text-align: right;">0.078218</td><td style="text-align: right;">0.015369</td><td style="text-align: right;">0.023895</td></tr><tr><td style="text-align: right;">0.000190</td><td style="text-align: right;">0.045242</td><td style="text-align: right;">0.035628</td><td style="text-align: right;">0.137340</td><td style="text-align: right;">0.130618</td><td style="text-align: right;">0.088636</td><td style="text-align: right;">0.016210</td><td style="text-align: right;">0.018537</td><td style="text-align: right;">0.333528</td><td style="text-align: right;">0.194070</td></tr><tr><td style="text-align: right;">0.004862</td><td style="text-align: right;">0.012371</td><td style="text-align: right;">0.041892</td><td style="text-align: right;">0.122430</td><td style="text-align: right;">0.125837</td><td style="text-align: right;">0.269677</td><td style="text-align: right;">0.018179</td><td style="text-align: right;">0.358968</td><td style="text-align: right;">0.012865</td><td style="text-align: right;">0.032920</td></tr><tr><td style="text-align: right;">0.006669</td><td style="text-align: right;">0.098761</td><td style="text-align: right;">0.382126</td><td style="text-align: right;">0.017893</td><td style="text-align: right;">0.120851</td><td style="text-align: right;">0.217545</td><td style="text-align: right;">0.089815</td><td style="text-align: right;">0.030089</td><td style="text-align: right;">0.026949</td><td style="text-align: right;">0.009302</td></tr><tr><td style="text-align: right;">0.012878</td><td style="text-align: right;">0.015030</td><td style="text-align: right;">0.065365</td><td style="text-align: right;">0.117578</td><td style="text-align: right;">0.162297</td><td style="text-align: right;">0.431229</td><td style="text-align: right;">0.021971</td><td style="text-align: right;">0.143659</td><td style="text-align: right;">0.013894</td><td style="text-align: right;">0.016100</td></tr><tr><td style="text-align: right;">0.001585</td><td style="text-align: right;">0.006070</td><td style="text-align: right;">0.054747</td><td style="text-align: right;">0.126926</td><td style="text-align: right;">0.119225</td><td style="text-align: right;">0.148794</td><td style="text-align: right;">0.021286</td><td style="text-align: right;">0.439636</td><td style="text-align: right;">0.044732</td><td style="text-align: right;">0.036999</td></tr><tr><td style="text-align: right;">0.000141</td><td style="text-align: right;">0.046332</td><td style="text-align: right;">0.035392</td><td style="text-align: right;">0.139779</td><td style="text-align: right;">0.129806</td><td style="text-align: right;">0.085847</td><td style="text-align: right;">0.020899</td><td style="text-align: right;">0.017417</td><td style="text-align: right;">0.354895</td><td style="text-align: right;">0.169491</td></tr><tr><td style="text-align: right;">0.000926</td><td style="text-align: right;">0.038906</td><td style="text-align: right;">0.049223</td><td style="text-align: right;">0.126108</td><td style="text-align: right;">0.136867</td><td style="text-align: right;">0.049725</td><td style="text-align: right;">0.009717</td><td style="text-align: right;">0.083438</td><td style="text-align: right;">0.418867</td><td style="text-align: right;">0.086222</td></tr><tr><td style="text-align: right;">0.000485</td><td style="text-align: right;">0.044163</td><td style="text-align: right;">0.041502</td><td style="text-align: right;">0.133406</td><td style="text-align: right;">0.132873</td><td style="text-align: right;">0.070729</td><td style="text-align: right;">0.010908</td><td style="text-align: right;">0.044333</td><td style="text-align: right;">0.395390</td><td style="text-align: right;">0.126211</td></tr><tr><td style="text-align: right;">0.018880</td><td style="text-align: right;">0.256336</td><td style="text-align: right;">0.286996</td><td style="text-align: right;">0.012342</td><td style="text-align: right;">0.134300</td><td style="text-align: right;">0.150295</td><td style="text-align: right;">0.074690</td><td style="text-align: right;">0.038984</td><td style="text-align: right;">0.021542</td><td style="text-align: right;">0.005635</td></tr><tr><td style="text-align: right;">...</td><td style="text-align: right;">...</td><td style="text-align: right;">...</td><td style="text-align: right;">...</td><td style="text-align: right;">...</td><td style="text-align: right;">...</td><td style="text-align: right;">...</td><td style="text-align: right;">...</td><td style="text-align: right;">...</td><td style="text-align: right;">...</td></tr><tr><td style="text-align: right;">0.255868</td><td style="text-align: right;">0.391587</td><td style="text-align: right;">0.033532</td><td style="text-align: right;">0.138212</td><td style="text-align: right;">0.040708</td><td style="text-align: right;">0.101267</td><td style="text-align: right;">0.009670</td><td style="text-align: right;">0.017948</td><td style="text-align: right;">0.002227</td><td style="text-align: right;">0.008983</td></tr><tr><td style="text-align: right;">0.000078</td><td style="text-align: right;">0.035770</td><td style="text-align: right;">0.000738</td><td style="text-align: right;">0.357107</td><td style="text-align: right;">0.139102</td><td style="text-align: right;">0.076811</td><td style="text-align: right;">0.263179</td><td style="text-align: right;">0.038327</td><td style="text-align: right;">0.049657</td><td style="text-align: right;">0.039232</td></tr><tr><td style="text-align: right;">0.001019</td><td style="text-align: right;">0.027571</td><td style="text-align: right;">0.042595</td><td style="text-align: right;">0.122011</td><td style="text-align: right;">0.137382</td><td style="text-align: right;">0.083578</td><td style="text-align: right;">0.017821</td><td style="text-align: right;">0.048060</td><td style="text-align: right;">0.152180</td><td style="text-align: right;">0.367783</td></tr><tr><td style="text-align: right;">0.050686</td><td style="text-align: right;">0.387217</td><td style="text-align: right;">0.058863</td><td style="text-align: right;">0.119808</td><td style="text-align: right;">0.205655</td><td style="text-align: right;">0.072896</td><td style="text-align: right;">0.019189</td><td style="text-align: right;">0.071453</td><td style="text-align: right;">0.003911</td><td style="text-align: right;">0.010321</td></tr><tr><td style="text-align: right;">0.000683</td><td style="text-align: right;">0.021987</td><td style="text-align: right;">0.033932</td><td style="text-align: right;">0.136035</td><td style="text-align: right;">0.152493</td><td style="text-align: right;">0.079471</td><td style="text-align: right;">0.429060</td><td style="text-align: right;">0.070587</td><td style="text-align: right;">0.031374</td><td style="text-align: right;">0.044377</td></tr><tr><td style="text-align: right;">0.000352</td><td style="text-align: right;">0.003326</td><td style="text-align: right;">0.053941</td><td style="text-align: right;">0.117159</td><td style="text-align: right;">0.113851</td><td style="text-align: right;">0.125731</td><td style="text-align: right;">0.022785</td><td style="text-align: right;">0.487746</td><td style="text-align: right;">0.036286</td><td style="text-align: right;">0.038822</td></tr><tr><td style="text-align: right;">0.000724</td><td style="text-align: right;">0.021481</td><td style="text-align: right;">0.035777</td><td style="text-align: right;">0.129189</td><td style="text-align: right;">0.152426</td><td style="text-align: right;">0.079149</td><td style="text-align: right;">0.424320</td><td style="text-align: right;">0.071530</td><td style="text-align: right;">0.033108</td><td style="text-align: right;">0.052295</td></tr><tr><td style="text-align: right;">0.000513</td><td style="text-align: right;">0.025270</td><td style="text-align: right;">0.024930</td><td style="text-align: right;">0.178402</td><td style="text-align: right;">0.150222</td><td style="text-align: right;">0.080508</td><td style="text-align: right;">0.399261</td><td style="text-align: right;">0.062214</td><td style="text-align: right;">0.035220</td><td style="text-align: right;">0.043459</td></tr><tr><td style="text-align: right;">0.000670</td><td style="text-align: right;">0.022160</td><td style="text-align: right;">0.033479</td><td style="text-align: right;">0.138869</td><td style="text-align: right;">0.152686</td><td style="text-align: right;">0.079395</td><td style="text-align: right;">0.435171</td><td style="text-align: right;">0.070602</td><td style="text-align: right;">0.030791</td><td style="text-align: right;">0.036178</td></tr><tr><td style="text-align: right;">0.000767</td><td style="text-align: right;">0.021881</td><td style="text-align: right;">0.037006</td><td style="text-align: right;">0.127669</td><td style="text-align: right;">0.149734</td><td style="text-align: right;">0.079379</td><td style="text-align: right;">0.366558</td><td style="text-align: right;">0.067744</td><td style="text-align: right;">0.045960</td><td style="text-align: right;">0.103302</td></tr></table></div><p>1776 cells x 10 lineages</p></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_macrostates()
g.plot_macrostates(discrete=True)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_31_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_31_0.png"/>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_31_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_31_1.png"/>
</div>
</div>
<p>When calculating 10 macrostates we notice that 2, 7 and 8 have clear macrostates with many around them.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_macrostates(same_plot=False, ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_33_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_33_0.png"/>
</div>
</div>
<p>We will now define the macrostates 2, 7 and 8 as our terminal states.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.set_terminal_states_from_macrostates(["2", "7", "8"])
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_terminal_states()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_36_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_36_0.png"/>
</div>
</div>
</section>
<section id="Calculating-absorption-probabilities">
<h2>Calculating absorption probabilities<a class="headerlink" href="#Calculating-absorption-probabilities" title="Permalink to this headline">#</a></h2>
<p>As a next step we will calculate the <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/api/cellrank.tl.estimators.GPCCA.compute_absorption_probabilities.html#cellrank.tl.estimators.GPCCA.compute_absorption_probabilities">absorption probabilities</a>.</p>
<p>For each visit, this computes the probability of being absorbed in any of the terminal_states. In particular, this corresponds to the probability that a random walk initialized in transient visit will reach any visit from a fixed transient state before reaching a visit from any other transient state.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_absorption_probabilities()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Unable to import petsc4py. For installation, please refer to: https://petsc4py.readthedocs.io/en/stable/install.html.
Defaulting to `'gmres'` solver.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "c2715f5a55384677921fd208b345ca20", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_absorption_probabilities()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_40_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_40_0.png"/>
</div>
</div>
<p>We can also calculate a pseudotime which is required for all patient visits to reach these states.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_absorption_probabilities(time_to_absorption="all")
g.absorption_times
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "289a021cb0bc4edf9fdd72ecb8f61b83", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8492970a7a5644dd86bb8f9b79eb34f7", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<div class="table-wrapper"><table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>2, 7, 8 mean</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>0.000000</td>
</tr>
<tr>
<th>1</th>
<td>692.089905</td>
</tr>
<tr>
<th>2</th>
<td>780.979614</td>
</tr>
<tr>
<th>3</th>
<td>315.616699</td>
</tr>
<tr>
<th>4</th>
<td>787.579163</td>
</tr>
<tr>
<th>...</th>
<td>...</td>
</tr>
<tr>
<th>1771</th>
<td>104.752121</td>
</tr>
<tr>
<th>1772</th>
<td>151.598206</td>
</tr>
<tr>
<th>1773</th>
<td>113.601944</td>
</tr>
<tr>
<th>1774</th>
<td>208.376373</td>
</tr>
<tr>
<th>1775</th>
<td>976.670654</td>
</tr>
</tbody>
</table></div>
<p>1776 rows × 1 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs["mean_time_to_absorption"] = g.absorption_times["2, 7, 8 mean"]
scv.pl.scatter(adata, color="mean_time_to_absorption")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_43_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_43_0.png"/>
</div>
</div>
<p>As a next step we want to determine the major features driving these transitions and terminal states.</p>
<section id="Extracting-lineage-drivers">
<h3>Extracting lineage drivers<a class="headerlink" href="#Extracting-lineage-drivers" title="Permalink to this headline">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_lineage_drivers()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<div class="table-wrapper"><table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>2_corr</th>
<th>2_pval</th>
<th>2_qval</th>
<th>2_ci_low</th>
<th>2_ci_high</th>
<th>7_corr</th>
<th>7_pval</th>
<th>7_qval</th>
<th>7_ci_low</th>
<th>7_ci_high</th>
<th>8_corr</th>
<th>8_pval</th>
<th>8_qval</th>
<th>8_ci_low</th>
<th>8_ci_high</th>
</tr>
</thead>
<tbody>
<tr>
<th>age</th>
<td>0.198010</td>
<td>2.932980e-17</td>
<td>NaN</td>
<td>0.152904</td>
<td>0.242292</td>
<td>0.074037</td>
<td>1.789044e-03</td>
<td>NaN</td>
<td>0.027618</td>
<td>0.120137</td>
<td>-0.226681</td>
<td>2.638689e-22</td>
<td>NaN</td>
<td>-0.270345</td>
<td>-0.182088</td>
</tr>
<tr>
<th>day_28_flg</th>
<td>0.171674</td>
<td>2.856733e-13</td>
<td>NaN</td>
<td>0.126168</td>
<td>0.216459</td>
<td>0.000759</td>
<td>9.744894e-01</td>
<td>NaN</td>
<td>-0.045756</td>
<td>0.047271</td>
<td>-0.154059</td>
<td>6.192002e-11</td>
<td>NaN</td>
<td>-0.199146</td>
<td>-0.108322</td>
</tr>
<tr>
<th>bun_first</th>
<td>0.157266</td>
<td>2.431946e-11</td>
<td>NaN</td>
<td>0.111568</td>
<td>0.202300</td>
<td>-0.024382</td>
<td>3.044819e-01</td>
<td>NaN</td>
<td>-0.070816</td>
<td>0.022156</td>
<td>-0.124396</td>
<td>1.400347e-07</td>
<td>NaN</td>
<td>-0.169926</td>
<td>-0.078335</td>
</tr>
<tr>
<th>chf_flg</th>
<td>0.153662</td>
<td>6.942518e-11</td>
<td>NaN</td>
<td>0.107919</td>
<td>0.198755</td>
<td>0.027422</td>
<td>2.481194e-01</td>
<td>NaN</td>
<td>-0.019116</td>
<td>0.073841</td>
<td>-0.155832</td>
<td>3.702895e-11</td>
<td>NaN</td>
<td>-0.200889</td>
<td>-0.110116</td>
</tr>
<tr>
<th>hosp_exp_flg</th>
<td>0.150377</td>
<td>1.767352e-10</td>
<td>NaN</td>
<td>0.104595</td>
<td>0.195523</td>
<td>0.002225</td>
<td>9.253618e-01</td>
<td>NaN</td>
<td>-0.044293</td>
<td>0.048733</td>
<td>-0.135987</td>
<td>8.328122e-09</td>
<td>NaN</td>
<td>-0.181353</td>
<td>-0.090042</td>
</tr>
<tr>
<th>copd_flg</th>
<td>0.118738</td>
<td>5.076657e-07</td>
<td>NaN</td>
<td>0.072625</td>
<td>0.164344</td>
<td>-0.010733</td>
<td>6.513101e-01</td>
<td>NaN</td>
<td>-0.057218</td>
<td>0.035799</td>
<td>-0.099055</td>
<td>2.856140e-05</td>
<td>NaN</td>
<td>-0.144901</td>
<td>-0.052785</td>
</tr>
<tr>
<th>afib_flg</th>
<td>0.118707</td>
<td>5.111579e-07</td>
<td>NaN</td>
<td>0.072594</td>
<td>0.164313</td>
<td>0.052288</td>
<td>2.754537e-02</td>
<td>NaN</td>
<td>0.005788</td>
<td>0.098562</td>
<td>-0.141180</td>
<td>2.168544e-09</td>
<td>NaN</td>
<td>-0.186469</td>
<td>-0.095292</td>
</tr>
<tr>
<th>hour_icu_intime</th>
<td>0.118526</td>
<td>5.321471e-07</td>
<td>NaN</td>
<td>0.072412</td>
<td>0.164135</td>
<td>0.184702</td>
<td>3.617098e-15</td>
<td>NaN</td>
<td>0.139386</td>
<td>0.229246</td>
<td>-0.229639</td>
<td>7.207144e-23</td>
<td>NaN</td>
<td>-0.273234</td>
<td>-0.185102</td>
</tr>
<tr>
<th>icu_exp_flg</th>
<td>0.114964</td>
<td>1.160405e-06</td>
<td>NaN</td>
<td>0.068818</td>
<td>0.160619</td>
<td>0.026165</td>
<td>2.704607e-01</td>
<td>NaN</td>
<td>-0.020373</td>
<td>0.072591</td>
<td>-0.120329</td>
<td>3.555081e-07</td>
<td>NaN</td>
<td>-0.165914</td>
<td>-0.074231</td>
</tr>
<tr>
<th>tco2_first</th>
<td>0.114953</td>
<td>1.163079e-06</td>
<td>NaN</td>
<td>0.068807</td>
<td>0.160608</td>
<td>0.023737</td>
<td>3.174620e-01</td>
<td>NaN</td>
<td>-0.022802</td>
<td>0.070173</td>
<td>-0.118734</td>
<td>5.081482e-07</td>
<td>NaN</td>
<td>-0.164340</td>
<td>-0.072621</td>
</tr>
<tr>
<th>resp_flg</th>
<td>0.104780</td>
<td>9.500759e-06</td>
<td>NaN</td>
<td>0.058552</td>
<td>0.150560</td>
<td>-0.016368</td>
<td>4.906492e-01</td>
<td>NaN</td>
<td>-0.062834</td>
<td>0.030168</td>
<td>-0.082799</td>
<td>4.750987e-04</td>
<td>NaN</td>
<td>-0.128816</td>
<td>-0.036426</td>
</tr>
<tr>
<th>creatinine_first</th>
<td>0.086735</td>
<td>2.508866e-04</td>
<td>NaN</td>
<td>0.040384</td>
<td>0.132713</td>
<td>-0.005053</td>
<td>8.314958e-01</td>
<td>NaN</td>
<td>-0.051555</td>
<td>0.041470</td>
<td>-0.074229</td>
<td>1.740237e-03</td>
<td>NaN</td>
<td>-0.120327</td>
<td>-0.027811</td>
</tr>
<tr>
<th>potassium_first</th>
<td>0.084717</td>
<td>3.492533e-04</td>
<td>NaN</td>
<td>0.038354</td>
<td>0.130715</td>
<td>0.004300</td>
<td>8.563360e-01</td>
<td>NaN</td>
<td>-0.042223</td>
<td>0.050803</td>
<td>-0.078682</td>
<td>9.003750e-04</td>
<td>NaN</td>
<td>-0.124739</td>
<td>-0.032287</td>
</tr>
<tr>
<th>stroke_flg</th>
<td>0.076318</td>
<td>1.282861e-03</td>
<td>NaN</td>
<td>0.029911</td>
<td>0.122397</td>
<td>0.103667</td>
<td>1.182249e-05</td>
<td>NaN</td>
<td>0.057430</td>
<td>0.149460</td>
<td>-0.137625</td>
<td>5.477792e-09</td>
<td>NaN</td>
<td>-0.182967</td>
<td>-0.091698</td>
</tr>
<tr>
<th>pco2_first</th>
<td>0.075803</td>
<td>1.384024e-03</td>
<td>NaN</td>
<td>0.029393</td>
<td>0.121887</td>
<td>-0.002868</td>
<td>9.038748e-01</td>
<td>NaN</td>
<td>-0.049375</td>
<td>0.043651</td>
<td>-0.065909</td>
<td>5.448075e-03</td>
<td>NaN</td>
<td>-0.112079</td>
<td>-0.019455</td>
</tr>
<tr>
<th>renal_flg</th>
<td>0.059138</td>
<td>1.266551e-02</td>
<td>NaN</td>
<td>0.012659</td>
<td>0.105362</td>
<td>0.002848</td>
<td>9.045331e-01</td>
<td>NaN</td>
<td>-0.043671</td>
<td>0.049355</td>
<td>-0.054814</td>
<td>2.086767e-02</td>
<td>NaN</td>
<td>-0.101070</td>
<td>-0.008322</td>
</tr>
<tr>
<th>platelet_first</th>
<td>0.031571</td>
<td>1.835897e-01</td>
<td>NaN</td>
<td>-0.014965</td>
<td>0.077970</td>
<td>0.018164</td>
<td>4.443304e-01</td>
<td>NaN</td>
<td>-0.028374</td>
<td>0.064623</td>
<td>-0.040404</td>
<td>8.871425e-02</td>
<td>NaN</td>
<td>-0.086755</td>
<td>0.006121</td>
</tr>
<tr>
<th>mal_flg</th>
<td>0.031065</td>
<td>1.907060e-01</td>
<td>NaN</td>
<td>-0.015471</td>
<td>0.077467</td>
<td>0.023017</td>
<td>3.323803e-01</td>
<td>NaN</td>
<td>-0.023522</td>
<td>0.069456</td>
<td>-0.043184</td>
<td>6.884053e-02</td>
<td>NaN</td>
<td>-0.089517</td>
<td>0.003337</td>
</tr>
<tr>
<th>cad_flg</th>
<td>0.029454</td>
<td>2.147627e-01</td>
<td>NaN</td>
<td>-0.017083</td>
<td>0.075864</td>
<td>0.037748</td>
<td>1.117878e-01</td>
<td>NaN</td>
<td>-0.008781</td>
<td>0.084114</td>
<td>-0.051604</td>
<td>2.964494e-02</td>
<td>NaN</td>
<td>-0.097882</td>
<td>-0.005102</td>
</tr>
<tr>
<th>temp_1st</th>
<td>0.023235</td>
<td>3.278209e-01</td>
<td>NaN</td>
<td>-0.023304</td>
<td>0.069673</td>
<td>-0.039728</td>
<td>9.418706e-02</td>
<td>NaN</td>
<td>-0.086083</td>
<td>0.006798</td>
<td>0.005777</td>
<td>8.078093e-01</td>
<td>NaN</td>
<td>-0.040748</td>
<td>0.052277</td>
</tr>
<tr>
<th>bmi</th>
<td>0.020368</td>
<td>3.910307e-01</td>
<td>NaN</td>
<td>-0.026171</td>
<td>0.066818</td>
<td>-0.031636</td>
<td>1.826791e-01</td>
<td>NaN</td>
<td>-0.078035</td>
<td>0.014899</td>
<td>0.002929</td>
<td>9.018525e-01</td>
<td>NaN</td>
<td>-0.043591</td>
<td>0.049436</td>
</tr>
<tr>
<th>ehrapycat_day_icu_intime</th>
<td>0.015719</td>
<td>5.080262e-01</td>
<td>NaN</td>
<td>-0.030818</td>
<td>0.062187</td>
<td>0.016911</td>
<td>4.763876e-01</td>
<td>NaN</td>
<td>-0.029626</td>
<td>0.063374</td>
<td>-0.025378</td>
<td>2.851418e-01</td>
<td>NaN</td>
<td>-0.071807</td>
<td>0.021160</td>
</tr>
<tr>
<th>icu_los_day</th>
<td>0.014934</td>
<td>5.294259e-01</td>
<td>NaN</td>
<td>-0.031601</td>
<td>0.061405</td>
<td>0.000388</td>
<td>9.869768e-01</td>
<td>NaN</td>
<td>-0.046127</td>
<td>0.046900</td>
<td>-0.013616</td>
<td>5.663999e-01</td>
<td>NaN</td>
<td>-0.060092</td>
<td>0.032919</td>
</tr>
<tr>
<th>sapsi_first</th>
<td>0.003322</td>
<td>8.887643e-01</td>
<td>NaN</td>
<td>-0.043199</td>
<td>0.049828</td>
<td>0.026311</td>
<td>2.678113e-01</td>
<td>NaN</td>
<td>-0.020228</td>
<td>0.072735</td>
<td>-0.020551</td>
<td>3.867760e-01</td>
<td>NaN</td>
<td>-0.067001</td>
<td>0.025987</td>
</tr>
<tr>
<th>day_icu_intime_num</th>
<td>-0.000906</td>
<td>9.695839e-01</td>
<td>NaN</td>
<td>-0.047417</td>
<td>0.045610</td>
<td>-0.013227</td>
<td>5.775484e-01</td>
<td>NaN</td>
<td>-0.059704</td>
<td>0.033307</td>
<td>0.009658</td>
<td>6.842281e-01</td>
<td>NaN</td>
<td>-0.036872</td>
<td>0.056147</td>
</tr>
<tr>
<th>map_1st</th>
<td>-0.001524</td>
<td>9.488330e-01</td>
<td>NaN</td>
<td>-0.048034</td>
<td>0.044993</td>
<td>0.014784</td>
<td>5.335757e-01</td>
<td>NaN</td>
<td>-0.031751</td>
<td>0.061256</td>
<td>-0.008524</td>
<td>7.196373e-01</td>
<td>NaN</td>
<td>-0.055016</td>
<td>0.038004</td>
</tr>
<tr>
<th>hospital_los_day</th>
<td>-0.001667</td>
<td>9.440521e-01</td>
<td>NaN</td>
<td>-0.048177</td>
<td>0.044850</td>
<td>-0.036409</td>
<td>1.250866e-01</td>
<td>NaN</td>
<td>-0.082783</td>
<td>0.010121</td>
<td>0.025844</td>
<td>2.763913e-01</td>
<td>NaN</td>
<td>-0.020694</td>
<td>0.072271</td>
</tr>
<tr>
<th>sodium_first</th>
<td>-0.006444</td>
<td>7.861307e-01</td>
<td>NaN</td>
<td>-0.052942</td>
<td>0.040082</td>
<td>0.009287</td>
<td>6.957527e-01</td>
<td>NaN</td>
<td>-0.037243</td>
<td>0.055777</td>
<td>-0.000458</td>
<td>9.846174e-01</td>
<td>NaN</td>
<td>-0.046971</td>
<td>0.046057</td>
</tr>
<tr>
<th>sofa_first</th>
<td>-0.007848</td>
<td>7.410585e-01</td>
<td>NaN</td>
<td>-0.054342</td>
<td>0.038680</td>
<td>-0.057878</td>
<td>1.469626e-02</td>
<td>NaN</td>
<td>-0.104111</td>
<td>-0.011395</td>
<td>0.045751</td>
<td>5.388327e-02</td>
<td>NaN</td>
<td>-0.000764</td>
<td>0.092068</td>
</tr>
<tr>
<th>weight_first</th>
<td>-0.014409</td>
<td>5.440132e-01</td>
<td>NaN</td>
<td>-0.060882</td>
<td>0.032126</td>
<td>-0.049309</td>
<td>3.771402e-02</td>
<td>NaN</td>
<td>-0.095604</td>
<td>-0.002802</td>
<td>0.045876</td>
<td>5.322974e-02</td>
<td>NaN</td>
<td>-0.000639</td>
<td>0.092193</td>
</tr>
<tr>
<th>abg_count</th>
<td>-0.018539</td>
<td>4.349753e-01</td>
<td>NaN</td>
<td>-0.064996</td>
<td>0.027999</td>
<td>-0.037325</td>
<td>1.158657e-01</td>
<td>NaN</td>
<td>-0.083693</td>
<td>0.009205</td>
<td>0.041559</td>
<td>7.995977e-02</td>
<td>NaN</td>
<td>-0.004965</td>
<td>0.087902</td>
</tr>
<tr>
<th>liver_flg</th>
<td>-0.036980</td>
<td>1.192707e-01</td>
<td>NaN</td>
<td>-0.083351</td>
<td>0.009550</td>
<td>-0.034970</td>
<td>1.407276e-01</td>
<td>NaN</td>
<td>-0.081351</td>
<td>0.011562</td>
<td>0.056496</td>
<td>1.724625e-02</td>
<td>NaN</td>
<td>0.010009</td>
<td>0.102740</td>
</tr>
<tr>
<th>hgb_first</th>
<td>-0.041921</td>
<td>7.735983e-02</td>
<td>NaN</td>
<td>-0.088263</td>
<td>0.004601</td>
<td>-0.024246</td>
<td>3.071895e-01</td>
<td>NaN</td>
<td>-0.070680</td>
<td>0.022293</td>
<td>0.053723</td>
<td>2.355561e-02</td>
<td>NaN</td>
<td>0.007228</td>
<td>0.099987</td>
</tr>
<tr>
<th>hr_1st</th>
<td>-0.053794</td>
<td>2.337226e-02</td>
<td>NaN</td>
<td>-0.100057</td>
<td>-0.007299</td>
<td>-0.071139</td>
<td>2.695349e-03</td>
<td>NaN</td>
<td>-0.117264</td>
<td>-0.024707</td>
<td>0.095724</td>
<td>5.276734e-05</td>
<td>NaN</td>
<td>0.049430</td>
<td>0.141607</td>
</tr>
<tr>
<th>chloride_first</th>
<td>-0.066152</td>
<td>5.278570e-03</td>
<td>NaN</td>
<td>-0.112320</td>
<td>-0.019699</td>
<td>0.014888</td>
<td>5.307002e-01</td>
<td>NaN</td>
<td>-0.031647</td>
<td>0.061359</td>
<td>0.049218</td>
<td>3.806874e-02</td>
<td>NaN</td>
<td>0.002711</td>
<td>0.095513</td>
</tr>
<tr>
<th>wbc_first</th>
<td>-0.073715</td>
<td>1.873614e-03</td>
<td>NaN</td>
<td>-0.119818</td>
<td>-0.027295</td>
<td>-0.020306</td>
<td>3.924725e-01</td>
<td>NaN</td>
<td>-0.066757</td>
<td>0.026232</td>
<td>0.079544</td>
<td>7.895188e-04</td>
<td>NaN</td>
<td>0.033153</td>
<td>0.125593</td>
</tr>
<tr>
<th>aline_flg</th>
<td>-0.074659</td>
<td>1.635241e-03</td>
<td>NaN</td>
<td>-0.120754</td>
<td>-0.028244</td>
<td>-0.040248</td>
<td>8.995817e-02</td>
<td>NaN</td>
<td>-0.086599</td>
<td>0.006278</td>
<td>0.093733</td>
<td>7.544736e-05</td>
<td>NaN</td>
<td>0.047426</td>
<td>0.139637</td>
</tr>
<tr>
<th>gender_num</th>
<td>-0.078282</td>
<td>9.566211e-04</td>
<td>NaN</td>
<td>-0.124343</td>
<td>-0.031885</td>
<td>-0.020364</td>
<td>3.911131e-01</td>
<td>NaN</td>
<td>-0.066815</td>
<td>0.026174</td>
<td>0.083668</td>
<td>4.135483e-04</td>
<td>NaN</td>
<td>0.037300</td>
<td>0.129677</td>
</tr>
<tr>
<th>spo2_1st</th>
<td>-0.080459</td>
<td>6.857679e-04</td>
<td>NaN</td>
<td>-0.126499</td>
<td>-0.034072</td>
<td>0.011439</td>
<td>6.300215e-01</td>
<td>NaN</td>
<td>-0.035093</td>
<td>0.057922</td>
<td>0.064339</td>
<td>6.670284e-03</td>
<td>NaN</td>
<td>0.017879</td>
<td>0.110522</td>
</tr>
<tr>
<th>po2_first</th>
<td>-0.085018</td>
<td>3.325744e-04</td>
<td>NaN</td>
<td>-0.131013</td>
<td>-0.038657</td>
<td>0.013725</td>
<td>5.632919e-01</td>
<td>NaN</td>
<td>-0.032809</td>
<td>0.060200</td>
<td>0.066893</td>
<td>4.789317e-03</td>
<td>NaN</td>
<td>0.020443</td>
<td>0.113055</td>
</tr>
<tr>
<th>ehrapycat_service_unit</th>
<td>-0.100333</td>
<td>2.245535e-05</td>
<td>NaN</td>
<td>-0.146164</td>
<td>-0.054071</td>
<td>0.000174</td>
<td>9.941479e-01</td>
<td>NaN</td>
<td>-0.046340</td>
<td>0.046687</td>
<td>0.089664</td>
<td>1.533117e-04</td>
<td>NaN</td>
<td>0.043331</td>
<td>0.135612</td>
</tr>
<tr>
<th>service_num</th>
<td>-0.121843</td>
<td>2.521916e-07</td>
<td>NaN</td>
<td>-0.167408</td>
<td>-0.075759</td>
<td>0.003443</td>
<td>8.847327e-01</td>
<td>NaN</td>
<td>-0.043078</td>
<td>0.049949</td>
<td>0.106723</td>
<td>6.453010e-06</td>
<td>NaN</td>
<td>0.060510</td>
<td>0.152480</td>
</tr>
<tr>
<th>mort_day_censored</th>
<td>-0.165865</td>
<td>1.799390e-12</td>
<td>NaN</td>
<td>-0.210753</td>
<td>-0.120279</td>
<td>-0.034629</td>
<td>1.446409e-01</td>
<td>NaN</td>
<td>-0.081013</td>
<td>0.011903</td>
<td>0.171520</td>
<td>3.002075e-13</td>
<td>NaN</td>
<td>0.126012</td>
<td>0.216308</td>
</tr>
<tr>
<th>censor_flg</th>
<td>-0.185205</td>
<td>3.034777e-15</td>
<td>NaN</td>
<td>-0.229739</td>
<td>-0.139896</td>
<td>-0.034846</td>
<td>1.421420e-01</td>
<td>NaN</td>
<td>-0.081228</td>
<td>0.011686</td>
<td>0.188962</td>
<td>8.052285e-16</td>
<td>NaN</td>
<td>0.143711</td>
<td>0.233424</td>
</tr>
<tr>
<th>iv_day_1</th>
<td>-0.662480</td>
<td>4.822694e-247</td>
<td>NaN</td>
<td>-0.687800</td>
<td>-0.635551</td>
<td>-0.297956</td>
<td>2.731907e-38</td>
<td>NaN</td>
<td>-0.339761</td>
<td>-0.254976</td>
<td>0.792124</td>
<td>0.000000e+00</td>
<td>NaN</td>
<td>0.774133</td>
<td>0.808836</td>
</tr>
<tr>
<th>sepsis_flg</th>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
</tr>
</tbody>
</table></div>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_lineage_drivers(lineage="2", ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_47_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_47_0.png"/>
</div>
</div>
<p>Cluster 2 seems to have a lot of patients with COPD, congestive heart failure (chf) and atrial fibrillation (afib) as a comorbidities.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_lineage_drivers(lineage="7", ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_49_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_49_0.png"/>
</div>
</div>
<p>Cluster 7 is primarily influenced by atrial fibrillation (afib), coronary artery disease (cad), congestive heart failure (chf) and the SAPS I score (sapsi).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_lineage_drivers(lineage="8", ncols=3)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_51_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_51_0.png"/>
</div>
</div>
<p>Cluster 8 is primarly defined as a terminal state by the heart rate (hr) and the IV input fluids (iv).</p>
<p>Although some of the features (e.g. congestive heart failure and age) overlap as drivers for death as a terminal state, all three clusters have unique origins of their macrostates.</p>
<p>We can also easily examine how likely it is that patients from a specific cluster end up in one of our three death dominated clusters.</p>
</section>
</section>
<section id="Exploring-cluster-fates">
<h2>Exploring cluster fates<a class="headerlink" href="#Exploring-cluster-fates" title="Permalink to this headline">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.cluster_fates(adata, mode="bar", cluster_key="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_56_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_56_0.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.cluster_fates(adata, mode="heatmap", cluster_key="leiden_0_5")
cr.pl.cluster_fates(adata, mode="clustermap", cluster_key="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_57_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_57_0.png"/>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_57_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_57_1.png"/>
</div>
</div>
<p>We can also leverage PAGA to visualize this information in a PAGA graph.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.tl.paga(adata, groups="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.cluster_fates(adata, mode="paga_pie", basis="umap", cluster_key="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: transitions_confidence not found, using connectivites instead.
WARNING: Invalid color key. Using grey instead.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/networkx/drawing/nx_pylab.py:717: MatplotlibDeprecationWarning: Passing *transOffset* without *offsets* has no effect. This behavior is deprecated since 3.5 and in 3.6, *transOffset* will begin having an effect regardless of *offsets*. In the meantime, if you wish to set *transOffset*, call collection.set_offset_transform(transOffset) explicitly.
  edge_collection = mpl.collections.LineCollection(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_60_2.png" src="../../_images/tutorials_notebooks_mimic_2_fate_60_2.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.cluster_fates(
    adata, mode="paga", legend_loc="on data", basis="umap", cluster_key="leiden_0_5"
)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: transitions_confidence not found, using connectivites instead.
WARNING: Invalid color key. Using grey instead.
WARNING: transitions_confidence not found, using connectivites instead.
WARNING: Invalid color key. Using grey instead.
WARNING: transitions_confidence not found, using connectivites instead.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/networkx/drawing/nx_pylab.py:717: MatplotlibDeprecationWarning: Passing *transOffset* without *offsets* has no effect. This behavior is deprecated since 3.5 and in 3.6, *transOffset* will begin having an effect regardless of *offsets*. In the meantime, if you wish to set *transOffset*, call collection.set_offset_transform(transOffset) explicitly.
  edge_collection = mpl.collections.LineCollection(
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/networkx/drawing/nx_pylab.py:717: MatplotlibDeprecationWarning: Passing *transOffset* without *offsets* has no effect. This behavior is deprecated since 3.5 and in 3.6, *transOffset* will begin having an effect regardless of *offsets*. In the meantime, if you wish to set *transOffset*, call collection.set_offset_transform(transOffset) explicitly.
  edge_collection = mpl.collections.LineCollection(
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/networkx/drawing/nx_pylab.py:717: MatplotlibDeprecationWarning: Passing *transOffset* without *offsets* has no effect. This behavior is deprecated since 3.5 and in 3.6, *transOffset* will begin having an effect regardless of *offsets*. In the meantime, if you wish to set *transOffset*, call collection.set_offset_transform(transOffset) explicitly.
  edge_collection = mpl.collections.LineCollection(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Invalid color key. Using grey instead.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_61_3.png" src="../../_images/tutorials_notebooks_mimic_2_fate_61_3.png"/>
</div>
</div>
<p>Finally, we can use a <a class="reference external" href="https://cellrank.readthedocs.io/en/stable/auto_examples/plotting/plot_circular_embedding.html">circular projection</a> to represent the same information.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.circular_projection(
    adata, keys=["leiden_0_5", "kl_divergence"], legend_loc="upper right"
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_63_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_63_0.png"/>
</div>
</div>
</section>
</section>
<section id="Determining-patient-fate-with-a-PseudotimeKernel">
<h1>Determining patient fate with a PseudotimeKernel<a class="headerlink" href="#Determining-patient-fate-with-a-PseudotimeKernel" title="Permalink to this headline">#</a></h1>
<p>We have now learnt that for example death dominated cluster 2 origins from cluster 0. Hence, we will now set a patient from cluster 0 as our root for our pseudotime calculation and will verify that we would indeed end up in cluster 2.</p>
<p>Using pseudotime to find terminal states is even more useful when the terminal states are unknown and only a root cluster is clear.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = ep.dt.mimic_2(encoded=True)
ep.pp.knn_impute(adata)
ep.pp.pca(adata)
ep.pp.neighbors(adata)
ep.tl.umap(adata)
ep.tl.leiden(adata, resolution=0.5, key_added="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "fca263fbcbc54d24bf560ae71f8d93c3", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:111: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  if ip and hasattr(ip, 'kernel') and hasattr(ip.kernel, '_parent_header'):
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:112: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  self.msg_id = ip.kernel._parent_header['header']['msg_id']
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="color: #808000; text-decoration-color: #808000; font-weight: bold">Quality control metrics missing. Calculating...</span>
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "379803146f914dd68fdd0922065e6416", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:111: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  if ip and hasattr(ip, 'kernel') and hasattr(ip.kernel, '_parent_header'):
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:112: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  self.msg_id = ip.kernel._parent_header['header']['msg_id']
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:111: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  if ip and hasattr(ip, 'kernel') and hasattr(ip.kernel, '_parent_header'):
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:112: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  self.msg_id = ip.kernel._parent_header['header']['msg_id']
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:111: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  if ip and hasattr(ip, 'kernel') and hasattr(ip.kernel, '_parent_header'):
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:112: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  self.msg_id = ip.kernel._parent_header['header']['msg_id']
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:111: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  if ip and hasattr(ip, 'kernel') and hasattr(ip.kernel, '_parent_header'):
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/ipywidgets/widgets/widget_outpu
t.py:112: DeprecationWarning: Kernel._parent_header is deprecated in ipykernel 6. Use
.get_parent()
  self.msg_id = ip.kernel._parent_header['header']['msg_id']
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1525: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  if (hasattr(numpy, value)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/numba/core/ir_utils.py:1526: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  and def_val == getattr(numpy, value)):
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pl.umap(adata, color=["leiden_0_5"], title="Leiden 0.5")
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing 'service_unit' as categorical
... storing 'day_icu_intime' as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_67_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_67_1.png"/>
</div>
</div>
<section id="Pseudotime-Kernels-and-stream-embeddings">
<h2>Pseudotime Kernels and stream embeddings<a class="headerlink" href="#Pseudotime-Kernels-and-stream-embeddings" title="Permalink to this headline">#</a></h2>
<p>The Pseudotime kernel computes direct transition probabilities based on a KNN graph and pseudotime.</p>
<p>The KNN graph contains information about the (undirected) connectivities among cells, reflecting their similarity. Pseudotime can be used to either remove edges that point against the direction of increasing pseudotime, or to downweight them.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.uns["iroot"] = np.flatnonzero(adata.obs["leiden_0_5"] == "0")[0]
ep.tl.dpt(adata)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pk = PseudotimeKernel(adata)
pk.compute_transition_matrix()
pk.compute_projection(basis="umap")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "d1cb3aa2ff0241b7b14a3a30a25aed80", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scv.pl.velocity_embedding_stream(adata, vkey="T_fwd", basis="umap", color="leiden_0_5")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_72_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_72_0.png"/>
</div>
</div>
<p>Now the arrows are much more clear and point into the direction of cluster 2 from cluster 0. Apparently transitions from cluster 11 to cluster 7 are possible, but also transitions via clusters 1, 6, 4, 3 and 8.</p>
</section>
<section id="Simulating-transitions-with-random-walks">
<h2>Simulating transitions with random walks<a class="headerlink" href="#Simulating-transitions-with-random-walks" title="Permalink to this headline">#</a></h2>
<p>Cellrank makes it easy to simulate the behavior of random walks from specific clusters. This allows us to not only visualize where the patients end up, but also roughly how many in which clusters after a defined number of iterations. We can either just start walking…</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pk.plot_random_walks(
    100,
    start_ixs={"leiden_0_5": "0"},
    max_iter=100,
    show_progress_bar=False,
    ixs_legend_loc="best",
    seed=42,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_76_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_76_0.png"/>
</div>
</div>
<p>… or set a number of required hits in one or more terminal clusters. Here, we require 50 hits in one of our three “death” clusters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pk.plot_random_walks(
    200,
    start_ixs={"leiden_0_5": "0"},
    stop_ixs={"leiden_0_5": ["8", "7", "2"]},
    max_iter=100,
    successive_hits=50,
    show_progress_bar=False,
    cmap="viridis",
    seed=42,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_78_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_78_0.png"/>
</div>
</div>
</section>
<section id="id1">
<h2>Determining macrostates and terminal states<a class="headerlink" href="#id1" title="Permalink to this headline">#</a></h2>
<p>Similarly to above when using our ConnectivityKernel we can compute macrostate memberships.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g = cr.tl.estimators.GPCCA(pk)
g.compute_schur()
g.compute_macrostates(cluster_key="leiden_0_5")
g.macrostates_memberships
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Unable to import `petsc4py` or `slepc4py`. Using `method='brandts'`
WARNING: For `method='brandts'`, dense matrix is required. Densifying
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div class="rendered_html" style="scoped"><div class="table-wrapper"><table class="dataframe"><tr><th style="background-color: #279e68; color: #ffffff; text-align: center; word-wrap: break-word; max-width: 100px">2</th><th style="background-color: #ffbb78; color: #000000; text-align: center; word-wrap: break-word; max-width: 100px">10</th></tr><tr><td style="text-align: right;">0.000000</td><td style="text-align: right;">1.000000</td></tr><tr><td style="text-align: right;">0.010048</td><td style="text-align: right;">0.989952</td></tr><tr><td style="text-align: right;">0.000000</td><td style="text-align: right;">1.000000</td></tr><tr><td style="text-align: right;">0.000000</td><td style="text-align: right;">1.000000</td></tr><tr><td style="text-align: right;">0.000000</td><td style="text-align: right;">1.000000</td></tr><tr><td style="text-align: right;">0.000001</td><td style="text-align: right;">0.999999</td></tr><tr><td style="text-align: right;">0.012068</td><td style="text-align: right;">0.987932</td></tr><tr><td style="text-align: right;">0.000319</td><td style="text-align: right;">0.999681</td></tr><tr><td style="text-align: right;">0.000856</td><td style="text-align: right;">0.999144</td></tr><tr><td style="text-align: right;">0.000000</td><td style="text-align: right;">1.000000</td></tr><tr><td style="text-align: right;">...</td><td style="text-align: right;">...</td></tr><tr><td style="text-align: right;">0.000000</td><td style="text-align: right;">1.000000</td></tr><tr><td style="text-align: right;">0.171079</td><td style="text-align: right;">0.828921</td></tr><tr><td style="text-align: right;">0.041941</td><td style="text-align: right;">0.958059</td></tr><tr><td style="text-align: right;">0.000000</td><td style="text-align: right;">1.000000</td></tr><tr><td style="text-align: right;">0.871996</td><td style="text-align: right;">0.128004</td></tr><tr><td style="text-align: right;">0.000000</td><td style="text-align: right;">1.000000</td></tr><tr><td style="text-align: right;">0.937809</td><td style="text-align: right;">0.062191</td></tr><tr><td style="text-align: right;">0.809891</td><td style="text-align: right;">0.190109</td></tr><tr><td style="text-align: right;">1.000000</td><td style="text-align: right;">0.000000</td></tr><tr><td style="text-align: right;">0.893366</td><td style="text-align: right;">0.106634</td></tr></table></div><p>1776 cells x 2 lineages</p></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_macrostates()
g.plot_macrostates(discrete=True)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_82_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_82_0.png"/>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_82_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_82_1.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.plot_macrostates(same_plot=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_83_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_83_0.png"/>
</div>
</div>
<p>As expected cluster 2 is a macrostate. Furthermore, cluster 10 is apparently a macrostates. This can likely be attributed to the fact that the people in cluster 10 are IV+.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ep.pl.umap(adata, color="iv_day_1")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_85_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_85_0.png"/>
</div>
</div>
</section>
<section id="id2">
<h2>Calculating absorption probabilities<a class="headerlink" href="#id2" title="Permalink to this headline">#</a></h2>
<p>Analogously we can compute terminal states, the absorption probabilities and the absorption times.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_terminal_states()
g.plot_terminal_states()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_88_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_88_0.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_absorption_probabilities(preconditioner="ilu", tol=1e-10)
g.plot_absorption_probabilities()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "842c5e8a7880480bba4319e0b82442ec", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_89_1.png" src="../../_images/tutorials_notebooks_mimic_2_fate_89_1.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_absorption_probabilities(
    time_to_absorption="all", preconditioner="ilu", tol=1e-10
)
g.absorption_times
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "b7aa3315226b4f2ea934a5f620bfe9b5", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "a420afacc1a74f898cfee0024ed41b7e", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<div class="table-wrapper"><table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>2, 10 mean</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>175.853287</td>
</tr>
<tr>
<th>1</th>
<td>367.299744</td>
</tr>
<tr>
<th>2</th>
<td>283.839600</td>
</tr>
<tr>
<th>3</th>
<td>142.796097</td>
</tr>
<tr>
<th>4</th>
<td>259.296387</td>
</tr>
<tr>
<th>...</th>
<td>...</td>
</tr>
<tr>
<th>1771</th>
<td>0.000000</td>
</tr>
<tr>
<th>1772</th>
<td>117.102867</td>
</tr>
<tr>
<th>1773</th>
<td>0.000000</td>
</tr>
<tr>
<th>1774</th>
<td>108.287277</td>
</tr>
<tr>
<th>1775</th>
<td>84.715179</td>
</tr>
</tbody>
</table></div>
<p>1776 rows × 1 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs["mean_time_to_absorption"] = g.absorption_times["2, 10 mean"]
scv.pl.scatter(adata, color="mean_time_to_absorption")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_91_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_91_0.png"/>
</div>
</div>
</section>
<section id="id3">
<h2>Extracting lineage drivers<a class="headerlink" href="#id3" title="Permalink to this headline">#</a></h2>
<p>Furthermore, just like above we can visualize the lineage drivers.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>g.compute_lineage_drivers()
g.plot_lineage_drivers(lineage="2")
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_94_0.png" src="../../_images/tutorials_notebooks_mimic_2_fate_94_0.png"/>
</div>
</div>
<p>The primary lineage drivers for cluster 2 are atrial fibrillation (afib), congestive heart failure (chf), blood urea nitrogen (bun) and COPD.</p>
</section>
<section id="Determining-feature-trends">
<h2>Determining feature trends<a class="headerlink" href="#Determining-feature-trends" title="Permalink to this headline">#</a></h2>
<p>cellrank uses <a class="reference external" href="https://cellrank.readthedocs.io/en/latest/api/cellrank.ul.models.GAM.html">Generalized Additive Models (GAMs)</a> to determine trends of features.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model = cr.ul.models.GAM(adata)
cr.pl.gene_trends(
    adata,
    model,
    ["iv_day_1"],
    time_key="dpt_pseudotime",
    show_progress_bar=False,
)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "df1eb2a8574e46e8943d14999fdc237e", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_98_2.png" src="../../_images/tutorials_notebooks_mimic_2_fate_98_2.png"/>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cr.pl.gene_trends(
    adata,
    model,
    ["iv_day_1"],
    same_plot=True,
    hide_cells=True,
    time_key="dpt_pseudotime",
    show_progress_bar=False,
)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "c12b2aab2da84f05927e8737cf0519cf", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:649: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  bases = (x &gt;= aug_knots[:-1]).astype(np.int) * \
/home/zeth/miniconda3/envs/ehrapy/lib/python3.8/site-packages/pygam/utils.py:650: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  (x &lt; aug_knots[1:]).astype(np.int)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_mimic_2_fate_99_2.png" src="../../_images/tutorials_notebooks_mimic_2_fate_99_2.png"/>
</div>
</div>
<p>As expected, the IV value is only marginally important for macro state 2, but even more important for macrostate 10. This is reflected in the required pseudotime to reach macrostate 10 with increasing IV values.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="diabetes_130.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Diabetes 130 Vignette</div>
              </div>
              <svg><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="mimic_2_introduction.html">
              <svg><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">MIMIC-II IAC Introduction</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021, Lukas Heumos, Theislab
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            Contents
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">MIMIC-II Patient Fate</a><ul>
<li><a class="reference internal" href="#Importing-ehrapy-and-setting-plotting-parameters">Importing ehrapy and setting plotting parameters</a></li>
<li><a class="reference internal" href="#Getting-and-preprocessing-the-MIMIC-II-dataset">Getting and preprocessing the MIMIC-II dataset</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Determining-patient-fate-using-a-ConnectivityKernel">Determining patient fate using a ConnectivityKernel</a><ul>
<li><a class="reference internal" href="#Connectivity-Kernels-and-stream-embeddings">Connectivity Kernels and stream embeddings</a></li>
<li><a class="reference internal" href="#Determining-macrostates-and-terminal-states">Determining macrostates and terminal states</a></li>
<li><a class="reference internal" href="#Calculating-absorption-probabilities">Calculating absorption probabilities</a><ul>
<li><a class="reference internal" href="#Extracting-lineage-drivers">Extracting lineage drivers</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Exploring-cluster-fates">Exploring cluster fates</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Determining-patient-fate-with-a-PseudotimeKernel">Determining patient fate with a PseudotimeKernel</a><ul>
<li><a class="reference internal" href="#Pseudotime-Kernels-and-stream-embeddings">Pseudotime Kernels and stream embeddings</a></li>
<li><a class="reference internal" href="#Simulating-transitions-with-random-walks">Simulating transitions with random walks</a></li>
<li><a class="reference internal" href="#id1">Determining macrostates and terminal states</a></li>
<li><a class="reference internal" href="#id2">Calculating absorption probabilities</a></li>
<li><a class="reference internal" href="#id3">Extracting lineage drivers</a></li>
<li><a class="reference internal" href="#Determining-feature-trends">Determining feature trends</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/scripts/furo.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/design-tabs.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>